// @firecms/schema_inference/dist/index.es.js
function parseReferenceString(value) {
  if (!value) return null;
  let database = void 0;
  let fullPath = value;
  if (value.includes(":::")) {
    const [dbName, pathPart] = value.split(":::");
    if (dbName && dbName !== "(default)") {
      database = dbName;
    }
    fullPath = pathPart;
  }
  if (!fullPath || !fullPath.includes("/")) {
    return null;
  }
  const path = fullPath.substring(0, fullPath.lastIndexOf("/"));
  return { path, database };
}
function looksLikeReference(value) {
  if (typeof value !== "string") return false;
  return parseReferenceString(value) !== null;
}
function findCommonInitialStringInPath(valuesCount) {
  if (!valuesCount) return void 0;
  function getPath(value) {
    let pathString;
    if (typeof value === "string") {
      pathString = value;
    } else if (value.path) {
      pathString = value.path;
    } else {
      console.warn("findCommonInitialStringInPath: value is not a string or document with path", value);
      return void 0;
    }
    if (!pathString) return void 0;
    if (pathString.includes(":::")) {
      const [, pathPart] = pathString.split(":::");
      pathString = pathPart;
    }
    return pathString;
  }
  const strings = valuesCount.values.map((v) => getPath(v)).filter((v) => !!v);
  const pathWithSlash = strings.find((s) => s.includes("/"));
  if (!pathWithSlash)
    return void 0;
  const searchedPath = pathWithSlash.substr(0, pathWithSlash.lastIndexOf("/"));
  const yep = valuesCount.values.filter((value) => {
    const path = getPath(value);
    if (!path) return false;
    return path.startsWith(searchedPath);
  }).length > valuesCount.values.length / 3 * 2;
  return yep ? searchedPath : void 0;
}
function removeInitialAndTrailingSlashes(s) {
  return removeInitialSlash(removeTrailingSlash(s));
}
function removeInitialSlash(s) {
  if (s.startsWith("/"))
    return s.slice(1);
  else return s;
}
function removeTrailingSlash(s) {
  if (s.endsWith("/"))
    return s.slice(0, -1);
  else return s;
}
function extractEnumFromValues(values) {
  if (!Array.isArray(values)) {
    return [];
  }
  const enumValues = values.map((value) => {
    if (typeof value === "string") {
      return {
        id: value,
        label: unslugify(value)
      };
    } else
      return null;
  }).filter(Boolean);
  enumValues.sort((a, b) => a.label.localeCompare(b.label));
  return enumValues;
}
function prettifyIdentifier(input) {
  if (!input) return "";
  let text = input;
  text = text.replace(/([a-z])([A-Z])|([A-Z])([A-Z][a-z])/g, "$1$3 $2$4");
  text = text.replace(/[_-]+/g, " ");
  const s = text.trim().replace(/\b\w/g, (char) => char.toUpperCase());
  console.log("Prettified identifier:", {
    input,
    s
  });
  return s;
}
function unslugify(slug) {
  if (!slug) return "";
  if (slug.includes("-") || slug.includes("_") || !slug.includes(" ")) {
    const result = slug.replace(/[-_]/g, " ");
    return result.replace(/\w\S*/g, function(txt) {
      return txt.charAt(0).toUpperCase() + txt.substr(1);
    }).trim();
  } else {
    return slug.trim();
  }
}
function resolveEnumValues(input) {
  if (Array.isArray(input)) {
    return input;
  } else if (typeof input === "object" && input !== null) {
    return Object.entries(input).map(([id, value]) => typeof value === "string" ? {
      id,
      label: value
    } : value);
  } else {
    return void 0;
  }
}
function mergeDeep(target, source, ignoreUndefined = false) {
  const targetIsObject = isObject(target);
  const output = targetIsObject ? { ...target } : target;
  if (targetIsObject && isObject(source)) {
    Object.keys(source).forEach((key) => {
      const sourceElement = source[key];
      if (ignoreUndefined && sourceElement === void 0) {
        return;
      }
      if (sourceElement instanceof Date) {
        Object.assign(output, { [key]: new Date(sourceElement.getTime()) });
      } else if (isPlainObject(sourceElement)) {
        if (!(key in target))
          Object.assign(output, { [key]: sourceElement });
        else if (isPlainObject(target[key]))
          output[key] = mergeDeep(target[key], sourceElement);
        else
          Object.assign(output, { [key]: sourceElement });
      } else if (isObject(sourceElement)) {
        Object.assign(output, { [key]: sourceElement });
      } else {
        Object.assign(output, { [key]: sourceElement });
      }
    });
  }
  return output;
}
function isObject(item) {
  return item && typeof item === "object" && !Array.isArray(item);
}
function isPlainObject(obj) {
  if (typeof obj !== "object" || obj === null || Array.isArray(obj)) {
    return false;
  }
  return Object.getPrototypeOf(obj) === Object.prototype;
}
const IMAGE_EXTENSIONS = [".jpg", ".jpeg", ".png", ".webp", ".gif", ".avif"];
const AUDIO_EXTENSIONS = [".mp3", ".ogg", ".opus", ".aac"];
const VIDEO_EXTENSIONS = [".avi", ".mp4"];
const emailRegEx = /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/;
function buildStringProperty({
  totalDocsCount,
  valuesResult
}) {
  let stringProperty = {
    dataType: "string"
  };
  if (valuesResult) {
    const totalEntriesCount = valuesResult.values.length;
    const totalValues = Array.from(valuesResult.valuesCount.keys()).length;
    const config = {};
    const probablyAURL = valuesResult.values.filter((value) => typeof value === "string" && value.toString().startsWith("http")).length > totalDocsCount / 3 * 2;
    if (probablyAURL) {
      config.url = true;
    }
    const probablyAnEmail = valuesResult.values.filter((value) => typeof value === "string" && emailRegEx.test(value)).length > totalDocsCount / 3 * 2;
    if (probablyAnEmail) {
      config.email = true;
    }
    const probablyUserIds = valuesResult.values.filter((value) => typeof value === "string" && value.length === 28 && !value.includes(" ")).length > totalDocsCount / 3 * 2;
    if (probablyUserIds)
      config.readOnly = true;
    if (!probablyAnEmail && !probablyAURL && !probablyUserIds && !probablyAURL && totalValues < totalEntriesCount / 3) {
      const enumValues = extractEnumFromValues(Array.from(valuesResult.valuesCount.keys()));
      if (Object.keys(enumValues).length > 1)
        config.enumValues = enumValues;
    }
    if (!probablyAnEmail && !probablyAURL && !probablyUserIds && !probablyAURL && !config.enumValues) {
      const fileType = probableFileType(valuesResult, totalDocsCount);
      if (fileType) {
        config.storage = {
          acceptedFiles: [fileType],
          storagePath: findCommonInitialStringInPath(valuesResult) ?? "/"
        };
      }
    }
    if (Object.keys(config).length > 0)
      stringProperty = {
        ...stringProperty,
        ...config,
        editable: true
      };
  }
  return stringProperty;
}
function probableFileType(valuesCount, totalDocsCount) {
  const probablyAnImage = valuesCount.values.filter((value) => typeof value === "string" && IMAGE_EXTENSIONS.some((extension) => value.toString().endsWith(extension))).length > totalDocsCount / 3 * 2;
  const probablyAudio = valuesCount.values.filter((value) => typeof value === "string" && AUDIO_EXTENSIONS.some((extension) => value.toString().endsWith(extension))).length > totalDocsCount / 3 * 2;
  const probablyVideo = valuesCount.values.filter((value) => typeof value === "string" && VIDEO_EXTENSIONS.some((extension) => value.toString().endsWith(extension))).length > totalDocsCount / 3 * 2;
  const fileType = probablyAnImage ? "image/*" : probablyAudio ? "audio/*" : probablyVideo ? "video/*" : false;
  return fileType;
}
function buildValidation({
  totalDocsCount,
  valuesResult
}) {
  if (valuesResult) {
    const totalEntriesCount = valuesResult.values.length;
    if (totalDocsCount === totalEntriesCount)
      return {
        required: true
      };
  }
  return void 0;
}
function buildReferenceProperty({
  totalDocsCount,
  valuesResult
}) {
  const property = {
    dataType: "reference",
    path: findCommonInitialStringInPath(valuesResult) ?? "!!!FIX_ME!!!",
    editable: true
  };
  return property;
}
async function buildEntityPropertiesFromData(data, getType) {
  const typesCount = {};
  const valuesCount = {};
  if (data) {
    data.forEach((entry) => {
      if (entry) {
        Object.entries(entry).forEach(([key, value]) => {
          if (key.startsWith("_")) return;
          increaseMapTypeCount(typesCount, key, value, getType);
          increaseValuesCount(valuesCount, key, value, getType);
        });
      }
    });
  }
  return buildPropertiesFromCount(data.length, typesCount, valuesCount);
}
function buildPropertyFromData(data, property, getType) {
  const typesCount = {};
  const valuesCount = {};
  if (data) {
    data.forEach((entry) => {
      increaseTypeCount(property.dataType, typesCount, entry, getType);
      increaseValuesCount(valuesCount, "inferred_prop", entry, getType);
    });
  }
  const enumValues = "enumValues" in property ? resolveEnumValues(property["enumValues"]) : void 0;
  if (enumValues) {
    const newEnumValues = extractEnumFromValues(Array.from(valuesCount["inferred_prop"].valuesCount.keys()));
    return {
      ...property,
      enumValues: [...newEnumValues, ...enumValues]
    };
  }
  const generatedProperty = buildPropertyFromCount(
    "inferred_prop",
    data.length,
    property.dataType,
    typesCount,
    valuesCount["inferred_prop"]
  );
  return mergeDeep(generatedProperty, property);
}
function buildPropertiesOrder(properties, propertiesOrder, priorityKeys) {
  const lowerCasePriorityKeys = (priorityKeys ?? []).map((key) => key.toLowerCase());
  function propOrder(s) {
    const k = s.toLowerCase();
    if (lowerCasePriorityKeys.includes(k)) return 4;
    if (k === "title" || k === "name") return 3;
    if (k.includes("title") || k.includes("name")) return 2;
    if (k.includes("image") || k.includes("picture")) return 1;
    return 0;
  }
  const keys = propertiesOrder ?? Object.keys(properties);
  keys.sort();
  keys.sort((a, b) => {
    return propOrder(b) - propOrder(a);
  });
  return keys;
}
function increaseTypeCount(type, typesCount, fieldValue, getType) {
  if (type === "map") {
    if (fieldValue) {
      let mapTypesCount = typesCount[type];
      if (!mapTypesCount) {
        mapTypesCount = {};
        typesCount[type] = mapTypesCount;
      }
      Object.entries(fieldValue).forEach(([key, value]) => {
        increaseMapTypeCount(mapTypesCount, key, value, getType);
      });
    }
  } else if (type === "array") {
    let arrayTypesCount = typesCount[type];
    if (!arrayTypesCount) {
      arrayTypesCount = {};
      typesCount[type] = arrayTypesCount;
    }
    if (fieldValue && Array.isArray(fieldValue) && fieldValue.length > 0) {
      const arrayType = getMostProbableTypeInArray(fieldValue, getType);
      if (arrayType === "map") {
        let mapTypesCount = arrayTypesCount[arrayType];
        if (!mapTypesCount) {
          mapTypesCount = {};
        }
        fieldValue.forEach((value) => {
          if (value && typeof value === "object" && !Array.isArray(value)) {
            Object.entries(value).forEach(
              ([key, v]) => increaseMapTypeCount(mapTypesCount, key, v, getType)
            );
          }
        });
        arrayTypesCount[arrayType] = mapTypesCount;
      } else {
        if (!arrayTypesCount[arrayType]) arrayTypesCount[arrayType] = 1;
        else arrayTypesCount[arrayType]++;
      }
    }
  } else {
    if (!typesCount[type]) typesCount[type] = 1;
    else typesCount[type]++;
  }
}
function increaseMapTypeCount(typesCountRecord, key, fieldValue, getType) {
  if (key.startsWith("_")) return;
  let typesCount = typesCountRecord[key];
  if (!typesCount) {
    typesCount = {};
    typesCountRecord[key] = typesCount;
  }
  if (fieldValue != null) {
    const type = getType(fieldValue);
    increaseTypeCount(type, typesCount, fieldValue, getType);
  }
}
function increaseValuesCount(typeValuesRecord, key, fieldValue, getType) {
  if (key.startsWith("_")) return;
  const dataType = getType(fieldValue);
  let valuesRecord = typeValuesRecord[key];
  if (!valuesRecord) {
    valuesRecord = {
      values: [],
      valuesCount: /* @__PURE__ */ new Map()
    };
    typeValuesRecord[key] = valuesRecord;
  }
  if (dataType === "map") {
    let mapValuesRecord = valuesRecord.map;
    if (!mapValuesRecord) {
      mapValuesRecord = {};
      valuesRecord.map = mapValuesRecord;
    }
    if (fieldValue)
      Object.entries(fieldValue).forEach(
        ([subKey, value]) => increaseValuesCount(mapValuesRecord, subKey, value, getType)
      );
  } else if (dataType === "array") {
    if (Array.isArray(fieldValue)) {
      fieldValue.forEach((value) => {
        valuesRecord.values.push(value);
        valuesRecord.valuesCount.set(value, (valuesRecord.valuesCount.get(value) ?? 0) + 1);
      });
    }
  } else {
    if (fieldValue !== null && fieldValue !== void 0) {
      valuesRecord.values.push(fieldValue);
      valuesRecord.valuesCount.set(fieldValue, (valuesRecord.valuesCount.get(fieldValue) ?? 0) + 1);
    }
  }
}
function getHighestTypesCount(typesCount) {
  let highestCount = 0;
  Object.entries(typesCount).forEach(([type, count]) => {
    let countValue = 0;
    if (type === "map") {
      countValue = getHighestRecordCount(count);
    } else if (type === "array") {
      countValue = getHighestTypesCount(count);
    } else {
      countValue = count;
    }
    if (countValue > highestCount) {
      highestCount = countValue;
    }
  });
  return highestCount;
}
function getHighestRecordCount(record) {
  return Object.entries(record).map(([key, typesCount]) => getHighestTypesCount(typesCount)).reduce((a, b) => Math.max(a, b), 0);
}
function getMostProbableType(typesCount) {
  let highestCount = -1;
  let probableType = "string";
  Object.entries(typesCount).forEach(([type, count]) => {
    let countValue;
    if (type === "map") {
      countValue = getHighestRecordCount(count);
    } else if (type === "array") {
      countValue = getHighestTypesCount(count);
    } else {
      countValue = count;
    }
    if (countValue > highestCount) {
      highestCount = countValue;
      probableType = type;
    }
  });
  return probableType;
}
function buildPropertyFromCount(key, totalDocsCount, mostProbableType, typesCount, valuesResult) {
  let title;
  if (key) {
    title = prettifyIdentifier(key);
  }
  let result = void 0;
  if (mostProbableType === "map") {
    const highVariability = checkTypesCountHighVariability(typesCount);
    if (highVariability) {
      result = {
        dataType: "map",
        name: title,
        keyValue: true,
        properties: {}
      };
    }
    const properties = buildPropertiesFromCount(
      totalDocsCount,
      typesCount.map,
      valuesResult ? valuesResult.mapValues : void 0
    );
    result = {
      dataType: "map",
      name: title,
      properties
    };
  } else if (mostProbableType === "array") {
    const arrayTypesCount = typesCount.array;
    const arrayMostProbableType = getMostProbableType(arrayTypesCount);
    const of = buildPropertyFromCount(
      key,
      totalDocsCount,
      arrayMostProbableType,
      arrayTypesCount,
      valuesResult
    );
    result = {
      dataType: "array",
      name: title,
      of
    };
  }
  if (!result) {
    const propertyProps = {
      totalDocsCount,
      valuesResult
    };
    if (mostProbableType === "string") {
      result = buildStringProperty(propertyProps);
    } else if (mostProbableType === "reference") {
      result = buildReferenceProperty(propertyProps);
    } else {
      result = {
        dataType: mostProbableType
      };
    }
    if (title) {
      result.name = title;
    }
    const validation = buildValidation(propertyProps);
    if (validation) {
      result.validation = validation;
    }
  }
  return {
    ...result,
    editable: true
  };
}
function buildPropertiesFromCount(totalDocsCount, typesCountRecord, valuesCountRecord) {
  const res = {};
  Object.entries(typesCountRecord).forEach(([key, typesCount]) => {
    const mostProbableType = getMostProbableType(typesCount);
    res[key] = buildPropertyFromCount(
      key,
      totalDocsCount,
      mostProbableType,
      typesCount,
      valuesCountRecord ? valuesCountRecord[key] : void 0
    );
  });
  return res;
}
function countMaxDocumentsUnder(typesCount) {
  let count = 0;
  Object.entries(typesCount).forEach(([type, value]) => {
    if (typeof value === "object") {
      count = Math.max(count, countMaxDocumentsUnder(value));
    } else {
      count = Math.max(count, value);
    }
  });
  return count;
}
function getMostProbableTypeInArray(array, getType) {
  const typesCount = {};
  array.forEach((value) => {
    increaseTypeCount(getType(value), typesCount, value, getType);
  });
  return getMostProbableType(typesCount);
}
function checkTypesCountHighVariability(typesCount) {
  const maxCount = countMaxDocumentsUnder(typesCount);
  let keysWithFewValues = 0;
  Object.entries(typesCount.map ?? {}).forEach(([key, value]) => {
    const count = countMaxDocumentsUnder(value);
    if (count < maxCount / 3) {
      keysWithFewValues++;
    }
  });
  return keysWithFewValues / Object.entries(typesCount.map ?? {}).length > 0.5;
}
function inferTypeFromValue(value) {
  if (value === null || value === void 0) return "string";
  if (typeof value === "string") return "string";
  if (typeof value === "number") return "number";
  if (typeof value === "boolean") return "boolean";
  if (Array.isArray(value)) return "array";
  if (typeof value === "object") return "map";
  return "string";
}
export {
  buildEntityPropertiesFromData,
  buildPropertiesOrder,
  buildPropertyFromData,
  extractEnumFromValues,
  findCommonInitialStringInPath,
  inferTypeFromValue,
  isObject,
  isPlainObject,
  looksLikeReference,
  mergeDeep,
  parseReferenceString,
  prettifyIdentifier,
  removeInitialAndTrailingSlashes,
  removeInitialSlash,
  removeTrailingSlash,
  resolveEnumValues,
  unslugify
};
