// @firecms/data_import/dist/index.umd.js
(function(global, factory) {
  typeof exports === "object" && typeof module !== "undefined" ? factory(exports, require("react-compiler-runtime"), require("react/jsx-runtime"), require("react"), require("@firecms/core"), require("@firecms/ui"), require("@firecms/schema_inference"), require("xlsx"), require("@firecms/formex")) : typeof define === "function" && define.amd ? define(["exports", "react-compiler-runtime", "react/jsx-runtime", "react", "@firecms/core", "@firecms/ui", "@firecms/schema_inference", "xlsx", "@firecms/formex"], factory) : (global = typeof globalThis !== "undefined" ? globalThis : global || self, factory(global["FireCMS data import/export"] = {}, global.reactCompilerRuntime, global.jsxRuntime, global.React, global.core, global.ui, global.schema_inference, global.XLSX, global.formex));
})(this, (function(exports2, reactCompilerRuntime, jsxRuntime, React, core, ui, schema_inference, XLSX, formex) {
  "use strict";
  function _interopNamespaceDefault(e) {
    const n = Object.create(null, { [Symbol.toStringTag]: { value: "Module" } });
    if (e) {
      for (const k in e) {
        if (k !== "default") {
          const d = Object.getOwnPropertyDescriptor(e, k);
          Object.defineProperty(n, k, d.get ? d : {
            enumerable: true,
            get: () => e[k]
          });
        }
      }
    }
    n.default = e;
    return Object.freeze(n);
  }
  const XLSX__namespace = /* @__PURE__ */ _interopNamespaceDefault(XLSX);
  const useImportConfig = () => {
    const $ = reactCompilerRuntime.c(18);
    const [inUse, setInUse] = React.useState(false);
    let t0;
    if ($[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
      t0 = {};
      $[0] = t0;
    } else {
      t0 = $[0];
    }
    const [defaultValues, setDefaultValues] = React.useState(t0);
    const [idColumn, setIdColumn] = React.useState();
    let t1;
    if ($[1] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
      t1 = [];
      $[1] = t1;
    } else {
      t1 = $[1];
    }
    const [importData, setImportData] = React.useState(t1);
    let t2;
    if ($[2] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
      t2 = [];
      $[2] = t2;
    } else {
      t2 = $[2];
    }
    const [entities, setEntities] = React.useState(t2);
    let t3;
    if ($[3] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
      t3 = {};
      $[3] = t3;
    } else {
      t3 = $[3];
    }
    const [headersMapping, setHeadersMapping] = React.useState(t3);
    let t4;
    if ($[4] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
      t4 = [];
      $[4] = t4;
    } else {
      t4 = $[4];
    }
    const [headingsOrder, setHeadingsOrder] = React.useState(t4);
    let t5;
    if ($[5] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
      t5 = {};
      $[5] = t5;
    } else {
      t5 = $[5];
    }
    const [originProperties, setOriginProperties] = React.useState(t5);
    let t6;
    if ($[6] !== headersMapping || $[7] !== headingsOrder) {
      t6 = (headingsOrder ?? []).length > 0 ? headingsOrder : Object.keys(headersMapping);
      $[6] = headersMapping;
      $[7] = headingsOrder;
      $[8] = t6;
    } else {
      t6 = $[8];
    }
    let t7;
    if ($[9] !== defaultValues || $[10] !== entities || $[11] !== headersMapping || $[12] !== idColumn || $[13] !== importData || $[14] !== inUse || $[15] !== originProperties || $[16] !== t6) {
      t7 = {
        inUse,
        setInUse,
        idColumn,
        setIdColumn,
        entities,
        setEntities,
        importData,
        setImportData,
        headingsOrder: t6,
        setHeadingsOrder,
        headersMapping,
        setHeadersMapping,
        originProperties,
        setOriginProperties,
        defaultValues,
        setDefaultValues
      };
      $[9] = defaultValues;
      $[10] = entities;
      $[11] = headersMapping;
      $[12] = idColumn;
      $[13] = importData;
      $[14] = inUse;
      $[15] = originProperties;
      $[16] = t6;
      $[17] = t7;
    } else {
      t7 = $[17];
    }
    return t7;
  };
  function getXLSXHeaders(sheet) {
    let header = 0;
    const hdr = [];
    const o = {};
    if (sheet == null || sheet["!ref"] == null) return [];
    const range = o.range !== void 0 ? o.range : sheet["!ref"];
    let r;
    if (o.header === 1) header = 1;
    else if (o.header === "A") header = 2;
    else if (Array.isArray(o.header)) header = 3;
    switch (typeof range) {
      case "string":
        r = safeDecodeRange(range);
        break;
      case "number":
        r = safeDecodeRange(sheet["!ref"]);
        r.s.r = range;
        break;
      default:
        r = range;
    }
    const rr = XLSX__namespace.utils.encode_row(r.s.r);
    const cols = new Array(r.e.c - r.s.c + 1);
    for (let C = r.s.c; C <= r.e.c; ++C) {
      cols[C] = XLSX__namespace.utils.encode_col(C);
      const val = sheet[cols[C] + rr];
      switch (header) {
        case 1:
          hdr.push(C);
          break;
        case 2:
          hdr.push(cols[C]);
          break;
        case 3:
          hdr.push(o.header[C - r.s.c]);
          break;
        default:
          if (val === void 0) continue;
          hdr.push(XLSX__namespace.utils.format_cell(val));
      }
    }
    return hdr;
  }
  function safeDecodeRange(range) {
    const o = {
      s: {
        c: 0,
        r: 0
      },
      e: {
        c: 0,
        r: 0
      }
    };
    let idx = 0;
    let i = 0;
    let cc = 0;
    const len = range.length;
    for (idx = 0; i < len; ++i) {
      if ((cc = range.charCodeAt(i) - 64) < 1 || cc > 26) break;
      idx = 26 * idx + cc;
    }
    o.s.c = --idx;
    for (idx = 0; i < len; ++i) {
      if ((cc = range.charCodeAt(i) - 48) < 0 || cc > 9) break;
      idx = 10 * idx + cc;
    }
    o.s.r = --idx;
    if (i === len || range.charCodeAt(++i) === 58) {
      o.e.c = o.s.c;
      o.e.r = o.s.r;
      return o;
    }
    for (idx = 0; i !== len; ++i) {
      if ((cc = range.charCodeAt(i) - 64) < 1 || cc > 26) break;
      idx = 26 * idx + cc;
    }
    o.e.c = --idx;
    for (idx = 0; i !== len; ++i) {
      if ((cc = range.charCodeAt(i) - 48) < 0 || cc > 9) break;
      idx = 10 * idx + cc;
    }
    o.e.r = --idx;
    return o;
  }
  function convertFileToJson(file) {
    return new Promise((resolve, reject) => {
      if (file.type === "application/json") {
        console.debug("Converting JSON file to JSON", file.name);
        const reader = new FileReader();
        reader.onload = function(e) {
          try {
            const data = e.target?.result;
            const jsonData = JSON.parse(data);
            if (!Array.isArray(jsonData)) {
              reject(new Error("JSON file should contain an array of objects"));
            } else {
              const propertiesOrder = jsonData.length > 0 ? Object.keys(jsonData[0]) : [];
              resolve({
                data: jsonData,
                propertiesOrder
              });
            }
          } catch (e2) {
            console.error("Error parsing JSON file", e2);
            reject(e2);
          }
        };
        reader.readAsText(file);
      } else {
        console.debug("Converting Excel file to JSON", file.name);
        const reader = new FileReader();
        reader.onload = function(e) {
          const data = new Uint8Array(e.target?.result);
          const workbook = XLSX__namespace.read(data, {
            type: "array",
            codepage: 65001,
            cellDates: true
          });
          const worksheetName = workbook.SheetNames[0];
          const worksheet = workbook.Sheets[worksheetName];
          const parsedData = XLSX__namespace.utils.sheet_to_json(worksheet);
          const headers = getXLSXHeaders(worksheet);
          const cleanedData = parsedData.map(mapJsonParse);
          const jsonData = cleanedData.map(unflattenObject);
          resolve({
            data: jsonData,
            propertiesOrder: headers
          });
        };
        reader.readAsArrayBuffer(file);
      }
    });
  }
  function mapJsonParse(obj) {
    return Object.keys(obj).reduce((acc, key) => {
      try {
        acc[key] = JSON.parse(obj[key]);
      } catch (e) {
        acc[key] = obj[key];
      }
      return acc;
    }, {});
  }
  function unflattenObject(flatObj) {
    return Object.keys(flatObj).reduce((nestedObj, key) => {
      let currentObj = nestedObj;
      const keyParts = key.split(".");
      keyParts.forEach((keyPart, i) => {
        if (/^[\w]+\[\d+\]$/.test(keyPart)) {
          const mainPropertyName = keyPart.slice(0, keyPart.indexOf("["));
          const index = parseInt(keyPart.slice(keyPart.indexOf("[") + 1, keyPart.indexOf("]")));
          if (!currentObj[mainPropertyName]) {
            currentObj[mainPropertyName] = [];
          }
          if (i !== keyParts.length - 1) {
            currentObj[mainPropertyName][index] = currentObj[mainPropertyName][index] || {};
            currentObj = currentObj[mainPropertyName][index];
          } else {
            currentObj[mainPropertyName][index] = flatObj[key];
          }
        } else if (i !== keyParts.length - 1) {
          currentObj[keyPart] = currentObj[keyPart] || {};
          currentObj = currentObj[keyPart];
        } else {
          currentObj[keyPart] = flatObj[key];
        }
      });
      return nestedObj;
    }, {});
  }
  function convertDataToEntity(authController, navigation, data, idColumn, headersMapping, properties, path, defaultValues) {
    const flatObject = flattenEntry(data);
    if (idColumn) delete flatObject[idColumn];
    const mappedKeysObject = Object.entries(flatObject).map(([key, value]) => {
      const mappedKey = formex.getIn(headersMapping, key) ?? key;
      const mappedProperty = core.getPropertyInPath(properties, mappedKey);
      if (!mappedProperty) {
        return {};
      }
      const processedValue = processValueMapping(authController, value, navigation, mappedProperty);
      return {
        [mappedKey]: processedValue
      };
    }).reduce((acc, curr) => ({
      ...acc,
      ...curr
    }), {});
    const values = core.mergeDeep(defaultValues ?? {}, unflattenObject(mappedKeysObject));
    let id = idColumn ? data[idColumn] : void 0;
    if (typeof id === "string") {
      id = id.trim();
    } else if (typeof id === "number") {
      id = id.toString();
    } else if (typeof id === "boolean") {
      id = id.toString();
    } else if (id instanceof Date) {
      id = id.toISOString();
    } else if (id && "toString" in id) {
      id = id.toString();
    }
    return {
      id,
      values,
      path
    };
  }
  function flattenEntry(obj, parent = "") {
    return Object.keys(obj).reduce((acc, key) => {
      const prefixedKey = parent ? `${parent}.${key}` : key;
      if (typeof obj[key] === "object" && !(obj[key] instanceof Date) && obj[key] !== null && !Array.isArray(obj[key])) {
        Object.assign(acc, flattenEntry(obj[key], prefixedKey));
      } else {
        acc[prefixedKey] = obj[key];
      }
      return acc;
    }, {});
  }
  function processValueMapping(authController, value, navigation, property) {
    if (value === null) return null;
    if (property === void 0) return value;
    const usedProperty = core.resolveProperty({
      propertyOrBuilder: property,
      authController
    });
    if (usedProperty === null) return value;
    const from = schema_inference.inferTypeFromValue(value);
    const to = usedProperty.dataType;
    if (from === "array" && to === "array" && Array.isArray(value) && usedProperty.of && !core.isPropertyBuilder(usedProperty.of)) {
      return value.map((v) => processValueMapping(authController, v, navigation, usedProperty.of));
    } else if (from === "string" && to === "number" && typeof value === "string") {
      return Number(value);
    } else if (from === "string" && to === "array" && typeof value === "string" && usedProperty.of && !core.isPropertyBuilder(usedProperty.of)) {
      return value.split(",").map((v) => processValueMapping(authController, v, usedProperty.of));
    } else if (from === "string" && to === "boolean") {
      return value === "true";
    } else if (from === "number" && to === "boolean") {
      return value === 1;
    } else if (from === "boolean" && to === "number") {
      return value ? 1 : 0;
    } else if (from === "boolean" && to === "string") {
      return value ? "true" : "false";
    } else if (from === "number" && to === "string" && typeof value === "number") {
      return value.toString();
    } else if (from === "string" && to === "array" && typeof value === "string") {
      return value.split(",").map((v) => v.trim());
    } else if (from === "string" && to === "date" && typeof value === "string") {
      try {
        return new Date(value);
      } catch (e) {
        return value;
      }
    } else if (from === "date" && to === "string") {
      return value instanceof Date && value.toISOString();
    } else if (from === "number" && to === "date" && typeof value === "number") {
      try {
        return new Date(value);
      } catch (e) {
        return value;
      }
    } else if (from === "string" && to === "reference" && typeof value === "string") {
      let databaseId = void 0;
      let referencePath = value;
      if (value.includes(":::")) {
        const [dbName, pathPart] = value.split(":::");
        if (dbName && dbName !== "(default)") {
          databaseId = dbName;
        }
        referencePath = pathPart;
      }
      const path = referencePath.split("/").slice(0, -1).join("/");
      const entityId = referencePath.split("/").slice(-1)[0];
      if (databaseId === void 0) {
        const targetCollection = navigation.getCollectionById(path);
        databaseId = targetCollection?.databaseId;
      }
      return new core.EntityReference(entityId, path, databaseId);
    } else if (from === to) {
      return value;
    } else if (from === "array" && to === "string" && Array.isArray(value)) {
      return value.join(",");
    }
    return value;
  }
  function getInferenceType(value) {
    if (typeof value === "number") return "number";
    else if (typeof value === "string") return "string";
    else if (typeof value === "boolean") return "boolean";
    else if (value instanceof Date) return "date";
    else if (Array.isArray(value)) return "array";
    return "map";
  }
  function DataNewPropertiesMapping(t0) {
    const $ = reactCompilerRuntime.c(35);
    const {
      importConfig,
      destinationProperties,
      buildPropertyView
    } = t0;
    const headersMapping = importConfig.headersMapping;
    const headingsOrder = importConfig.headingsOrder;
    const idColumn = importConfig.idColumn;
    const originProperties = importConfig.originProperties;
    let t1;
    if ($[0] !== importConfig) {
      t1 = (value) => importConfig.setIdColumn(value ?? void 0);
      $[0] = importConfig;
      $[1] = t1;
    } else {
      t1 = $[1];
    }
    let t2;
    if ($[2] !== headersMapping || $[3] !== idColumn || $[4] !== t1) {
      t2 = /* @__PURE__ */ jsxRuntime.jsx(IdSelectField, { idColumn, headersMapping, onChange: t1 });
      $[2] = headersMapping;
      $[3] = idColumn;
      $[4] = t1;
      $[5] = t2;
    } else {
      t2 = $[5];
    }
    let t3;
    if ($[6] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
      t3 = /* @__PURE__ */ jsxRuntime.jsx("div", { className: "h-4" });
      $[6] = t3;
    } else {
      t3 = $[6];
    }
    let t4;
    if ($[7] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
      t4 = {
        tableLayout: "fixed"
      };
      $[7] = t4;
    } else {
      t4 = $[7];
    }
    let t5;
    let t6;
    if ($[8] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
      t5 = /* @__PURE__ */ jsxRuntime.jsx(ui.TableCell, { header: true, style: {
        width: "20%"
      }, children: "Column in file" });
      t6 = /* @__PURE__ */ jsxRuntime.jsx(ui.TableCell, { header: true });
      $[8] = t5;
      $[9] = t6;
    } else {
      t5 = $[8];
      t6 = $[9];
    }
    let t7;
    if ($[10] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
      t7 = /* @__PURE__ */ jsxRuntime.jsxs(ui.TableHeader, { children: [
        t5,
        t6,
        /* @__PURE__ */ jsxRuntime.jsx(ui.TableCell, { header: true, style: {
          width: "75%"
        }, children: "Map to Property" })
      ] });
      $[10] = t7;
    } else {
      t7 = $[10];
    }
    let t8;
    if ($[11] !== buildPropertyView || $[12] !== destinationProperties || $[13] !== headersMapping || $[14] !== headingsOrder || $[15] !== idColumn || $[16] !== originProperties) {
      t8 = destinationProperties && headingsOrder.map((importKey) => {
        const mappedKey = headersMapping[importKey];
        const propertyKey = headersMapping[importKey];
        const property = mappedKey ? core.getPropertyInPath(destinationProperties, mappedKey) : null;
        const originProperty = core.getPropertyInPath(originProperties, importKey);
        const originDataType = originProperty ? originProperty.dataType === "array" && typeof originProperty.of === "object" ? `${originProperty.dataType} - ${originProperty.of.dataType}` : originProperty.dataType : void 0;
        return /* @__PURE__ */ jsxRuntime.jsxs(ui.TableRow, { style: {
          height: "90px"
        }, children: [
          /* @__PURE__ */ jsxRuntime.jsxs(ui.TableCell, { style: {
            width: "20%"
          }, children: [
            /* @__PURE__ */ jsxRuntime.jsx(ui.Typography, { variant: "body2", children: importKey }),
            originProperty && /* @__PURE__ */ jsxRuntime.jsx(ui.Typography, { variant: "caption", color: "secondary", children: originDataType })
          ] }),
          /* @__PURE__ */ jsxRuntime.jsx(ui.TableCell, { children: /* @__PURE__ */ jsxRuntime.jsx(ui.ChevronRightIcon, {}) }),
          /* @__PURE__ */ jsxRuntime.jsx(ui.TableCell, { className: importKey === idColumn ? "text-center" : void 0, style: {
            width: "75%"
          }, children: buildPropertyView?.({
            isIdColumn: importKey === idColumn,
            property,
            propertyKey,
            importKey
          }) })
        ] }, importKey);
      });
      $[11] = buildPropertyView;
      $[12] = destinationProperties;
      $[13] = headersMapping;
      $[14] = headingsOrder;
      $[15] = idColumn;
      $[16] = originProperties;
      $[17] = t8;
    } else {
      t8 = $[17];
    }
    let t9;
    if ($[18] !== t8) {
      t9 = /* @__PURE__ */ jsxRuntime.jsxs(ui.Table, { style: t4, children: [
        t7,
        /* @__PURE__ */ jsxRuntime.jsx(ui.TableBody, { children: t8 })
      ] });
      $[18] = t8;
      $[19] = t9;
    } else {
      t9 = $[19];
    }
    let t10;
    if ($[20] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
      t10 = /* @__PURE__ */ jsxRuntime.jsx("div", { className: "text-sm text-surface-accent-500 dark:text-surface-accent-300 font-medium ml-3.5 mb-1", children: "You can select a default value for unmapped columns and empty values:" });
      $[20] = t10;
    } else {
      t10 = $[20];
    }
    let t11;
    if ($[21] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
      t11 = {
        tableLayout: "fixed"
      };
      $[21] = t11;
    } else {
      t11 = $[21];
    }
    let t12;
    let t13;
    if ($[22] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
      t12 = /* @__PURE__ */ jsxRuntime.jsx(ui.TableCell, { header: true, style: {
        width: "30%"
      }, children: "Property" });
      t13 = /* @__PURE__ */ jsxRuntime.jsx(ui.TableCell, { header: true });
      $[22] = t12;
      $[23] = t13;
    } else {
      t12 = $[22];
      t13 = $[23];
    }
    let t14;
    if ($[24] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
      t14 = /* @__PURE__ */ jsxRuntime.jsxs(ui.TableHeader, { children: [
        t12,
        t13,
        /* @__PURE__ */ jsxRuntime.jsx(ui.TableCell, { header: true, style: {
          width: "65%"
        }, children: "Default value" })
      ] });
      $[24] = t14;
    } else {
      t14 = $[24];
    }
    let t15;
    if ($[25] !== destinationProperties || $[26] !== idColumn || $[27] !== importConfig) {
      t15 = destinationProperties && getAllPropertyKeys(destinationProperties).map((key) => {
        const property_0 = core.getPropertyInPath(destinationProperties, key);
        if (typeof property_0 !== "object" || property_0 === null) {
          return null;
        }
        if (!["number", "string", "boolean", "map"].includes(property_0.dataType)) {
          return null;
        }
        return /* @__PURE__ */ jsxRuntime.jsxs(ui.TableRow, { style: {
          height: "70px"
        }, children: [
          /* @__PURE__ */ jsxRuntime.jsx(ui.TableCell, { style: {
            width: "20%"
          }, children: /* @__PURE__ */ jsxRuntime.jsx(ui.Typography, { variant: "body2", children: key }) }),
          /* @__PURE__ */ jsxRuntime.jsx(ui.TableCell, { children: /* @__PURE__ */ jsxRuntime.jsx(ui.ChevronRightIcon, {}) }),
          /* @__PURE__ */ jsxRuntime.jsx(ui.TableCell, { className: key === idColumn ? "text-center" : void 0, style: {
            width: "75%"
          }, children: /* @__PURE__ */ jsxRuntime.jsx(DefaultValuesField, { property: property_0, defaultValue: formex.getIn(importConfig.defaultValues, key), onValueChange: (value_0) => {
            const newValues = formex.setIn(importConfig.defaultValues, key, value_0);
            importConfig.setDefaultValues(newValues);
          } }) })
        ] }, key);
      });
      $[25] = destinationProperties;
      $[26] = idColumn;
      $[27] = importConfig;
      $[28] = t15;
    } else {
      t15 = $[28];
    }
    let t16;
    if ($[29] !== t15) {
      t16 = /* @__PURE__ */ jsxRuntime.jsxs(ui.ExpandablePanel, { title: "Default values", initiallyExpanded: false, innerClassName: "p-4 mt-4", children: [
        t10,
        /* @__PURE__ */ jsxRuntime.jsxs(ui.Table, { style: t11, children: [
          t14,
          /* @__PURE__ */ jsxRuntime.jsx(ui.TableBody, { children: t15 })
        ] })
      ] });
      $[29] = t15;
      $[30] = t16;
    } else {
      t16 = $[30];
    }
    let t17;
    if ($[31] !== t16 || $[32] !== t2 || $[33] !== t9) {
      t17 = /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
        t2,
        t3,
        t9,
        t16
      ] });
      $[31] = t16;
      $[32] = t2;
      $[33] = t9;
      $[34] = t17;
    } else {
      t17 = $[34];
    }
    return t17;
  }
  function getAllPropertyKeys(properties, currentKey) {
    return Object.entries(properties).reduce((acc, [key, property]) => {
      const accumulatedKey = currentKey ? `${currentKey}.${key}` : key;
      if (typeof property !== "function" && property.dataType === "map" && property.properties) {
        const childProperties = getAllPropertyKeys(property.properties, accumulatedKey);
        return [...acc, ...childProperties];
      }
      return [...acc, accumulatedKey];
    }, []);
  }
  function IdSelectField(t0) {
    const $ = reactCompilerRuntime.c(9);
    const {
      idColumn,
      headersMapping,
      onChange
    } = t0;
    const t1 = idColumn ?? "";
    let t2;
    if ($[0] !== onChange) {
      t2 = (event) => {
        const value = event.target.value;
        onChange(value === "__none__" ? null : value);
      };
      $[0] = onChange;
      $[1] = t2;
    } else {
      t2 = $[1];
    }
    let t3;
    if ($[2] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
      t3 = /* @__PURE__ */ jsxRuntime.jsx(ui.SelectItem, { value: "__none__", children: "Autogenerate ID" });
      $[2] = t3;
    } else {
      t3 = $[2];
    }
    let t4;
    if ($[3] !== headersMapping) {
      t4 = Object.entries(headersMapping).map(_temp2$1);
      $[3] = headersMapping;
      $[4] = t4;
    } else {
      t4 = $[4];
    }
    let t5;
    if ($[5] !== t1 || $[6] !== t2 || $[7] !== t4) {
      t5 = /* @__PURE__ */ jsxRuntime.jsx("div", { children: /* @__PURE__ */ jsxRuntime.jsxs(ui.Select, { size: "medium", value: t1, fullWidth: true, onChange: t2, placeholder: "Autogenerate ID", renderValue: _temp$1, label: "Column that will be used as ID for each document", children: [
        t3,
        t4
      ] }) });
      $[5] = t1;
      $[6] = t2;
      $[7] = t4;
      $[8] = t5;
    } else {
      t5 = $[8];
    }
    return t5;
  }
  function _temp2$1(t0) {
    const [key] = t0;
    return /* @__PURE__ */ jsxRuntime.jsx(ui.SelectItem, { value: key, children: key }, key);
  }
  function _temp$1(value_0) {
    return /* @__PURE__ */ jsxRuntime.jsx(ui.Typography, { variant: "body2", children: value_0 !== "__none__" ? value_0 : "Autogenerate ID" });
  }
  function DefaultValuesField(t0) {
    const $ = reactCompilerRuntime.c(22);
    const {
      property,
      onValueChange,
      defaultValue
    } = t0;
    if (property.dataType === "string") {
      const t1 = defaultValue ?? "";
      let t2;
      if ($[0] !== onValueChange) {
        t2 = (event) => onValueChange(event.target.value);
        $[0] = onValueChange;
        $[1] = t2;
      } else {
        t2 = $[1];
      }
      let t3;
      if ($[2] !== t1 || $[3] !== t2) {
        t3 = /* @__PURE__ */ jsxRuntime.jsx(ui.TextField, { size: "medium", placeholder: "Default value", value: t1, onChange: t2 });
        $[2] = t1;
        $[3] = t2;
        $[4] = t3;
      } else {
        t3 = $[4];
      }
      return t3;
    } else {
      if (property.dataType === "number") {
        const t1 = defaultValue ?? "";
        let t2;
        if ($[5] !== onValueChange) {
          t2 = (event_0) => onValueChange(event_0.target.value);
          $[5] = onValueChange;
          $[6] = t2;
        } else {
          t2 = $[6];
        }
        let t3;
        if ($[7] !== t1 || $[8] !== t2) {
          t3 = /* @__PURE__ */ jsxRuntime.jsx(ui.TextField, { size: "medium", type: "number", value: t1, placeholder: "Default value", onChange: t2 });
          $[7] = t1;
          $[8] = t2;
          $[9] = t3;
        } else {
          t3 = $[9];
        }
        return t3;
      } else {
        if (property.dataType === "boolean") {
          const t1 = defaultValue ?? null;
          let t2;
          if ($[10] !== onValueChange) {
            t2 = (v) => onValueChange(v === null ? void 0 : v);
            $[10] = onValueChange;
            $[11] = t2;
          } else {
            t2 = $[11];
          }
          const t3 = defaultValue === void 0 ? "Do not set value" : defaultValue === true ? "Set value to true" : "Set value to false";
          let t4;
          if ($[12] !== t1 || $[13] !== t2 || $[14] !== t3) {
            t4 = /* @__PURE__ */ jsxRuntime.jsx(ui.BooleanSwitchWithLabel, { value: t1, allowIndeterminate: true, size: "small", onValueChange: t2, label: t3 });
            $[12] = t1;
            $[13] = t2;
            $[14] = t3;
            $[15] = t4;
          } else {
            t4 = $[15];
          }
          return t4;
        } else {
          if (property.dataType === "date") {
            const t1 = property.mode ?? "date";
            const t2 = defaultValue ?? void 0;
            let t3;
            if ($[16] !== onValueChange) {
              t3 = (dateValue) => {
                onValueChange(dateValue);
              };
              $[16] = onValueChange;
              $[17] = t3;
            } else {
              t3 = $[17];
            }
            let t4;
            if ($[18] !== t1 || $[19] !== t2 || $[20] !== t3) {
              t4 = /* @__PURE__ */ jsxRuntime.jsx(ui.DateTimeField, { mode: t1, size: "medium", value: t2, onChange: t3, clearable: true });
              $[18] = t1;
              $[19] = t2;
              $[20] = t3;
              $[21] = t4;
            } else {
              t4 = $[21];
            }
            return t4;
          }
        }
      }
    }
    return null;
  }
  function ImportFileUpload(t0) {
    const $ = reactCompilerRuntime.c(5);
    const {
      onDataAdded
    } = t0;
    const snackbarController = core.useSnackbarController();
    let t1;
    if ($[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
      t1 = {
        "text/*": [".csv", ".xls", ".xlsx"],
        "application/vnd.ms-excel": [".xls", ".xlsx"],
        "application/msexcel": [".xls", ".xlsx"],
        "application/vnd.ms-office": [".xls", ".xlsx"],
        "application/xls": [".xls", ".xlsx"],
        "application/x-xls": [".xls", ".xlsx"],
        "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet": [".xls", ".xlsx"],
        "application/json": [".json"]
      };
      $[0] = t1;
    } else {
      t1 = $[0];
    }
    let t2;
    if ($[1] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
      t2 = /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
        /* @__PURE__ */ jsxRuntime.jsx(ui.UploadIcon, {}),
        "Drag and drop a file here or click to upload"
      ] });
      $[1] = t2;
    } else {
      t2 = $[1];
    }
    let t3;
    if ($[2] !== onDataAdded || $[3] !== snackbarController) {
      t3 = /* @__PURE__ */ jsxRuntime.jsx(ui.FileUpload, { accept: t1, preventDropOnDocument: true, size: "small", maxFiles: 1, uploadDescription: t2, onFilesAdded: (files) => {
        if (files.length > 0) {
          convertFileToJson(files[0]).then((t4) => {
            const {
              data,
              propertiesOrder
            } = t4;
            onDataAdded(data, propertiesOrder);
          }).catch((error) => {
            console.error("Error parsing file", error);
            snackbarController.open({
              type: "error",
              message: error.message
            });
          });
        }
      } });
      $[2] = onDataAdded;
      $[3] = snackbarController;
      $[4] = t3;
    } else {
      t3 = $[4];
    }
    return t3;
  }
  function ImportNewPropertyFieldPreview(t0) {
    const $ = reactCompilerRuntime.c(22);
    const {
      propertyKey,
      property,
      onEditClick,
      includeName: t1,
      onPropertyNameChanged,
      propertyTypeView
    } = t0;
    const includeName = t1 === void 0 ? true : t1;
    const {
      propertyConfigs
    } = core.useCustomizationController();
    let t2;
    if ($[0] !== property || $[1] !== propertyConfigs) {
      t2 = property ? core.getFieldConfig(property, propertyConfigs) : null;
      $[0] = property;
      $[1] = propertyConfigs;
      $[2] = t2;
    } else {
      t2 = $[2];
    }
    const widget = t2;
    let t3;
    if ($[3] !== propertyTypeView || $[4] !== widget) {
      t3 = propertyTypeView ?? /* @__PURE__ */ jsxRuntime.jsx(core.PropertyConfigBadge, { propertyConfig: widget ?? void 0 });
      $[3] = propertyTypeView;
      $[4] = widget;
      $[5] = t3;
    } else {
      t3 = $[5];
    }
    let t4;
    if ($[6] !== t3) {
      t4 = /* @__PURE__ */ jsxRuntime.jsx("div", { className: "mx-4", children: t3 });
      $[6] = t3;
      $[7] = t4;
    } else {
      t4 = $[7];
    }
    let t5;
    if ($[8] !== includeName || $[9] !== onPropertyNameChanged || $[10] !== property?.name || $[11] !== propertyKey) {
      t5 = includeName && /* @__PURE__ */ jsxRuntime.jsx(ui.TextField, { size: "medium", className: "text-base grow", value: property?.name ?? "", onChange: (e) => {
        if (onPropertyNameChanged && propertyKey) {
          onPropertyNameChanged(propertyKey, e.target.value);
        }
      } });
      $[8] = includeName;
      $[9] = onPropertyNameChanged;
      $[10] = property?.name;
      $[11] = propertyKey;
      $[12] = t5;
    } else {
      t5 = $[12];
    }
    let t6;
    if ($[13] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
      t6 = /* @__PURE__ */ jsxRuntime.jsx(ui.EditIcon, { size: "small" });
      $[13] = t6;
    } else {
      t6 = $[13];
    }
    let t7;
    if ($[14] !== onEditClick) {
      t7 = /* @__PURE__ */ jsxRuntime.jsx(ui.IconButton, { onClick: onEditClick, size: "small", children: t6 });
      $[14] = onEditClick;
      $[15] = t7;
    } else {
      t7 = $[15];
    }
    let t8;
    if ($[16] !== t5 || $[17] !== t7) {
      t8 = /* @__PURE__ */ jsxRuntime.jsx("div", { className: "w-full flex flex-col grow", children: /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "flex flex-row items-center gap-2", children: [
        t5,
        t7
      ] }) });
      $[16] = t5;
      $[17] = t7;
      $[18] = t8;
    } else {
      t8 = $[18];
    }
    let t9;
    if ($[19] !== t4 || $[20] !== t8) {
      t9 = /* @__PURE__ */ jsxRuntime.jsx(core.ErrorBoundary, { children: /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "flex flex-row w-full items-center", children: [
        t4,
        t8
      ] }) });
      $[19] = t4;
      $[20] = t8;
      $[21] = t9;
    } else {
      t9 = $[21];
    }
    return t9;
  }
  function ImportSaveInProgress(t0) {
    const $ = reactCompilerRuntime.c(25);
    const {
      path,
      importConfig,
      collection,
      onImportSuccess
    } = t0;
    const [errorSaving, setErrorSaving] = React.useState(void 0);
    const dataSource = core.useDataSource();
    const savingRef = React.useRef(false);
    const [processedEntities, setProcessedEntities] = React.useState(0);
    let t1;
    if ($[0] !== collection || $[1] !== dataSource || $[2] !== importConfig || $[3] !== onImportSuccess || $[4] !== path) {
      t1 = function save2() {
        if (savingRef.current) {
          return;
        }
        savingRef.current = true;
        saveDataBatch(dataSource, collection, path, importConfig.entities, 0, 25, setProcessedEntities).then(() => {
          onImportSuccess(collection);
          savingRef.current = false;
        }).catch((e) => {
          setErrorSaving(e);
          savingRef.current = false;
        });
      };
      $[0] = collection;
      $[1] = dataSource;
      $[2] = importConfig;
      $[3] = onImportSuccess;
      $[4] = path;
      $[5] = t1;
    } else {
      t1 = $[5];
    }
    const save = t1;
    let t2;
    if ($[6] !== save) {
      t2 = () => {
        save();
      };
      $[6] = save;
      $[7] = t2;
    } else {
      t2 = $[7];
    }
    let t3;
    if ($[8] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
      t3 = [];
      $[8] = t3;
    } else {
      t3 = $[8];
    }
    React.useEffect(t2, t3);
    if (errorSaving) {
      let t42;
      if ($[9] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
        t42 = /* @__PURE__ */ jsxRuntime.jsx(ui.Typography, { variant: "h6", children: "Error saving data" });
        $[9] = t42;
      } else {
        t42 = $[9];
      }
      let t52;
      if ($[10] !== errorSaving.message) {
        t52 = /* @__PURE__ */ jsxRuntime.jsx(ui.Typography, { variant: "body2", color: "error", children: errorSaving.message });
        $[10] = errorSaving.message;
        $[11] = t52;
      } else {
        t52 = $[11];
      }
      let t62;
      if ($[12] !== save) {
        t62 = /* @__PURE__ */ jsxRuntime.jsx(ui.Button, { onClick: save, children: "Retry" });
        $[12] = save;
        $[13] = t62;
      } else {
        t62 = $[13];
      }
      let t72;
      if ($[14] !== t52 || $[15] !== t62) {
        t72 = /* @__PURE__ */ jsxRuntime.jsxs(ui.CenteredView, { className: "flex flex-col gap-4 items-center", children: [
          t42,
          t52,
          t62
        ] });
        $[14] = t52;
        $[15] = t62;
        $[16] = t72;
      } else {
        t72 = $[16];
      }
      return t72;
    }
    let t4;
    let t5;
    if ($[17] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
      t4 = /* @__PURE__ */ jsxRuntime.jsx(ui.CircularProgress, {});
      t5 = /* @__PURE__ */ jsxRuntime.jsx(ui.Typography, { variant: "h6", children: "Saving data" });
      $[17] = t4;
      $[18] = t5;
    } else {
      t4 = $[17];
      t5 = $[18];
    }
    let t6;
    if ($[19] !== importConfig.entities.length || $[20] !== processedEntities) {
      t6 = /* @__PURE__ */ jsxRuntime.jsxs(ui.Typography, { variant: "body2", children: [
        processedEntities,
        "/",
        importConfig.entities.length,
        " entities saved"
      ] });
      $[19] = importConfig.entities.length;
      $[20] = processedEntities;
      $[21] = t6;
    } else {
      t6 = $[21];
    }
    let t7;
    if ($[22] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
      t7 = /* @__PURE__ */ jsxRuntime.jsx(ui.Typography, { variant: "caption", children: "Do not close this tab or the import will be interrupted." });
      $[22] = t7;
    } else {
      t7 = $[22];
    }
    let t8;
    if ($[23] !== t6) {
      t8 = /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "flex flex-col gap-4 items-center", children: [
        t4,
        t5,
        t6,
        t7
      ] });
      $[23] = t6;
      $[24] = t8;
    } else {
      t8 = $[24];
    }
    return t8;
  }
  function saveDataBatch(dataSource, collection, path, data, offset = 0, batchSize = 25, onProgressUpdate) {
    console.debug("Saving imported data", offset, batchSize);
    const batch = data.slice(offset, offset + batchSize);
    return Promise.all(batch.map((d) => dataSource.saveEntity({
      path,
      values: d.values,
      entityId: d.id,
      collection,
      status: "new"
    }))).then(() => {
      if (offset + batchSize < data.length) {
        onProgressUpdate(offset + batchSize);
        return saveDataBatch(dataSource, collection, path, data, offset + batchSize, batchSize, onProgressUpdate);
      }
      onProgressUpdate(data.length);
      return Promise.resolve();
    });
  }
  function ImportCollectionAction(t0) {
    const $ = reactCompilerRuntime.c(79);
    const {
      collection,
      path,
      onAnalyticsEvent
    } = t0;
    const authController = core.useAuthController();
    const customizationController = core.useCustomizationController();
    const snackbarController = core.useSnackbarController();
    const [open, setOpen] = React.useState(false);
    const [step, setStep] = React.useState("initial");
    const importConfig = useImportConfig();
    let t1;
    if ($[0] !== onAnalyticsEvent) {
      t1 = () => {
        setOpen(true);
        onAnalyticsEvent?.("import_open");
        setStep("initial");
      };
      $[0] = onAnalyticsEvent;
      $[1] = t1;
    } else {
      t1 = $[1];
    }
    const handleClickOpen = t1;
    let t2;
    if ($[2] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
      t2 = () => {
        setOpen(false);
      };
      $[2] = t2;
    } else {
      t2 = $[2];
    }
    const handleClose = t2;
    let t3;
    if ($[3] !== onAnalyticsEvent) {
      t3 = () => {
        onAnalyticsEvent?.("import_mapping_complete");
        setStep("preview");
      };
      $[3] = onAnalyticsEvent;
      $[4] = t3;
    } else {
      t3 = $[4];
    }
    const onMappingComplete = t3;
    let t4;
    if ($[5] !== onAnalyticsEvent) {
      t4 = () => {
        onAnalyticsEvent?.("import_data_save");
        setStep("import_data_saving");
      };
      $[5] = onAnalyticsEvent;
      $[6] = t4;
    } else {
      t4 = $[6];
    }
    const onPreviewComplete = t4;
    let t5;
    if ($[7] !== collection.properties || $[8] !== importConfig || $[9] !== onAnalyticsEvent) {
      t5 = async (data) => {
        importConfig.setImportData(data);
        if (data.length > 0) {
          const originProperties = await schema_inference.buildEntityPropertiesFromData(data, getInferenceType);
          importConfig.setOriginProperties(originProperties);
          const headersMapping = buildHeadersMappingFromData(data, collection?.properties);
          importConfig.setHeadersMapping(headersMapping);
          const firstKey = Object.keys(headersMapping)?.[0];
          if (firstKey?.includes("id") || firstKey?.includes("key")) {
            importConfig.setIdColumn(firstKey);
          }
        }
        setTimeout(() => {
          onAnalyticsEvent?.("import_data_added");
          setStep("mapping");
        }, 100);
      };
      $[7] = collection.properties;
      $[8] = importConfig;
      $[9] = onAnalyticsEvent;
      $[10] = t5;
    } else {
      t5 = $[10];
    }
    const onDataAdded = t5;
    let properties;
    let resolvedCollection;
    let t6;
    if ($[11] !== authController || $[12] !== collection || $[13] !== customizationController.propertyConfigs || $[14] !== path) {
      resolvedCollection = core.resolveCollection({
        collection,
        path,
        propertyConfigs: customizationController.propertyConfigs,
        authController
      });
      properties = core.getPropertiesWithPropertiesOrder(resolvedCollection.properties, resolvedCollection.propertiesOrder);
      t6 = Object.entries(properties).flatMap(_temp);
      $[11] = authController;
      $[12] = collection;
      $[13] = customizationController.propertyConfigs;
      $[14] = path;
      $[15] = properties;
      $[16] = resolvedCollection;
      $[17] = t6;
    } else {
      properties = $[15];
      resolvedCollection = $[16];
      t6 = $[17];
    }
    const propertiesAndLevel = t6;
    let t7;
    if ($[18] !== resolvedCollection.properties || $[19] !== resolvedCollection.propertiesOrder) {
      t7 = resolvedCollection.propertiesOrder ?? Object.keys(resolvedCollection.properties);
      $[18] = resolvedCollection.properties;
      $[19] = resolvedCollection.propertiesOrder;
      $[20] = t7;
    } else {
      t7 = $[20];
    }
    const propertiesOrder = t7;
    if (collection.collectionGroup) {
      return null;
    }
    let t8;
    if ($[21] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
      t8 = /* @__PURE__ */ jsxRuntime.jsx(ui.UploadIcon, { size: "small" });
      $[21] = t8;
    } else {
      t8 = $[21];
    }
    let t9;
    if ($[22] !== handleClickOpen) {
      t9 = /* @__PURE__ */ jsxRuntime.jsx(ui.Tooltip, { title: "Import", asChild: true, children: /* @__PURE__ */ jsxRuntime.jsx(ui.IconButton, { size: "small", color: "primary", onClick: handleClickOpen, children: t8 }) });
      $[22] = handleClickOpen;
      $[23] = t9;
    } else {
      t9 = $[23];
    }
    const t10 = step !== "initial";
    const t11 = step !== "initial";
    const t12 = step === "initial" ? "lg" : "7xl";
    const t13 = step === "preview";
    let t14;
    if ($[24] !== t13) {
      t14 = /* @__PURE__ */ jsxRuntime.jsx(ui.DialogTitle, { variant: "h6", hidden: t13, children: "Import data" });
      $[24] = t13;
      $[25] = t14;
    } else {
      t14 = $[25];
    }
    const t15 = step === "preview";
    let t16;
    if ($[26] !== onDataAdded || $[27] !== step) {
      t16 = step === "initial" && /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
        /* @__PURE__ */ jsxRuntime.jsx(ui.Typography, { variant: "body2", children: "Upload a CSV, Excel or JSON file and map it to your existing schema" }),
        /* @__PURE__ */ jsxRuntime.jsx(ImportFileUpload, { onDataAdded })
      ] });
      $[26] = onDataAdded;
      $[27] = step;
      $[28] = t16;
    } else {
      t16 = $[28];
    }
    let t17;
    if ($[29] !== importConfig || $[30] !== onAnalyticsEvent || $[31] !== properties || $[32] !== propertiesAndLevel || $[33] !== step) {
      t17 = step === "mapping" && /* @__PURE__ */ jsxRuntime.jsx(jsxRuntime.Fragment, { children: /* @__PURE__ */ jsxRuntime.jsx(DataNewPropertiesMapping, { importConfig, destinationProperties: properties, buildPropertyView: (t182) => {
        const {
          isIdColumn,
          propertyKey,
          importKey
        } = t182;
        return /* @__PURE__ */ jsxRuntime.jsx(PropertyTreeSelect, { selectedPropertyKey: propertyKey ?? "", properties, propertiesAndLevel, isIdColumn, onIdSelected: () => {
          importConfig.setIdColumn(importKey);
        }, onPropertySelected: (newPropertyKey) => {
          onAnalyticsEvent?.("import_mapping_field_updated");
          const newHeadersMapping = Object.entries(importConfig.headersMapping).map((t192) => {
            const [currentImportKey, currentPropertyKey] = t192;
            if (currentPropertyKey === newPropertyKey) {
              return {
                [currentImportKey]: null
              };
            }
            if (currentImportKey === importKey) {
              return {
                [currentImportKey]: newPropertyKey
              };
            }
            return {
              [currentImportKey]: currentPropertyKey
            };
          }).reduce(_temp2, {});
          importConfig.setHeadersMapping(newHeadersMapping);
          if (newPropertyKey === importConfig.idColumn) {
            importConfig.setIdColumn(void 0);
          }
        } });
      } }) });
      $[29] = importConfig;
      $[30] = onAnalyticsEvent;
      $[31] = properties;
      $[32] = propertiesAndLevel;
      $[33] = step;
      $[34] = t17;
    } else {
      t17 = $[34];
    }
    let t18;
    if ($[35] !== importConfig || $[36] !== properties || $[37] !== propertiesOrder || $[38] !== step) {
      t18 = step === "preview" && /* @__PURE__ */ jsxRuntime.jsx(ImportDataPreview, { importConfig, properties, propertiesOrder });
      $[35] = importConfig;
      $[36] = properties;
      $[37] = propertiesOrder;
      $[38] = step;
      $[39] = t18;
    } else {
      t18 = $[39];
    }
    let t19;
    if ($[40] !== collection || $[41] !== importConfig || $[42] !== path || $[43] !== snackbarController || $[44] !== step) {
      t19 = step === "import_data_saving" && importConfig && /* @__PURE__ */ jsxRuntime.jsx(ImportSaveInProgress, { importConfig, collection, path, onImportSuccess: (importedCollection) => {
        handleClose();
        snackbarController.open({
          type: "info",
          message: "Data imported successfully"
        });
      } });
      $[40] = collection;
      $[41] = importConfig;
      $[42] = path;
      $[43] = snackbarController;
      $[44] = step;
      $[45] = t19;
    } else {
      t19 = $[45];
    }
    let t20;
    if ($[46] !== t15 || $[47] !== t16 || $[48] !== t17 || $[49] !== t18 || $[50] !== t19) {
      t20 = /* @__PURE__ */ jsxRuntime.jsxs(ui.DialogContent, { className: "h-full flex flex-col gap-4 my-4", fullHeight: t15, children: [
        t16,
        t17,
        t18,
        t19
      ] });
      $[46] = t15;
      $[47] = t16;
      $[48] = t17;
      $[49] = t18;
      $[50] = t19;
      $[51] = t20;
    } else {
      t20 = $[51];
    }
    let t21;
    if ($[52] !== step) {
      t21 = step === "mapping" && /* @__PURE__ */ jsxRuntime.jsx(ui.Button, { onClick: () => setStep("initial"), variant: "text", children: "Back" });
      $[52] = step;
      $[53] = t21;
    } else {
      t21 = $[53];
    }
    let t22;
    if ($[54] !== step) {
      t22 = step === "preview" && /* @__PURE__ */ jsxRuntime.jsx(ui.Button, { onClick: () => setStep("mapping"), variant: "text", children: "Back" });
      $[54] = step;
      $[55] = t22;
    } else {
      t22 = $[55];
    }
    let t23;
    if ($[56] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
      t23 = /* @__PURE__ */ jsxRuntime.jsx(ui.Button, { onClick: handleClose, variant: "text", children: "Cancel" });
      $[56] = t23;
    } else {
      t23 = $[56];
    }
    let t24;
    if ($[57] !== onMappingComplete || $[58] !== step) {
      t24 = step === "mapping" && /* @__PURE__ */ jsxRuntime.jsx(ui.Button, { variant: "filled", color: "primary", onClick: onMappingComplete, children: "Next" });
      $[57] = onMappingComplete;
      $[58] = step;
      $[59] = t24;
    } else {
      t24 = $[59];
    }
    let t25;
    if ($[60] !== onPreviewComplete || $[61] !== step) {
      t25 = step === "preview" && /* @__PURE__ */ jsxRuntime.jsx(ui.Button, { variant: "filled", color: "primary", onClick: onPreviewComplete, children: "Save data" });
      $[60] = onPreviewComplete;
      $[61] = step;
      $[62] = t25;
    } else {
      t25 = $[62];
    }
    let t26;
    if ($[63] !== t21 || $[64] !== t22 || $[65] !== t24 || $[66] !== t25) {
      t26 = /* @__PURE__ */ jsxRuntime.jsxs(ui.DialogActions, { children: [
        t21,
        t22,
        t23,
        t24,
        t25
      ] });
      $[63] = t21;
      $[64] = t22;
      $[65] = t24;
      $[66] = t25;
      $[67] = t26;
    } else {
      t26 = $[67];
    }
    let t27;
    if ($[68] !== open || $[69] !== t10 || $[70] !== t11 || $[71] !== t12 || $[72] !== t14 || $[73] !== t20 || $[74] !== t26) {
      t27 = /* @__PURE__ */ jsxRuntime.jsxs(ui.Dialog, { open, fullWidth: t10, fullHeight: t11, maxWidth: t12, children: [
        t14,
        t20,
        t26
      ] });
      $[68] = open;
      $[69] = t10;
      $[70] = t11;
      $[71] = t12;
      $[72] = t14;
      $[73] = t20;
      $[74] = t26;
      $[75] = t27;
    } else {
      t27 = $[75];
    }
    let t28;
    if ($[76] !== t27 || $[77] !== t9) {
      t28 = /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
        t9,
        t27
      ] });
      $[76] = t27;
      $[77] = t9;
      $[78] = t28;
    } else {
      t28 = $[78];
    }
    return t28;
  }
  function _temp2(acc, curr) {
    return {
      ...acc,
      ...curr
    };
  }
  function _temp(t0) {
    const [key, property] = t0;
    return getPropertiesAndLevel(key, property, 0);
  }
  const internalIDValue = "__internal_id__";
  function PropertyTreeSelect(t0) {
    const $ = reactCompilerRuntime.c(17);
    const {
      selectedPropertyKey,
      properties,
      onPropertySelected,
      onIdSelected,
      propertiesAndLevel,
      isIdColumn
    } = t0;
    let t1;
    if ($[0] !== properties || $[1] !== selectedPropertyKey) {
      t1 = selectedPropertyKey ? core.getPropertyInPath(properties, selectedPropertyKey) : null;
      $[0] = properties;
      $[1] = selectedPropertyKey;
      $[2] = t1;
    } else {
      t1 = $[2];
    }
    const selectedProperty = t1;
    let t2;
    if ($[3] !== selectedProperty) {
      t2 = (selectedPropertyKey_0) => {
        if (selectedPropertyKey_0 === internalIDValue) {
          return /* @__PURE__ */ jsxRuntime.jsx(ui.Typography, { variant: "body2", className: "p-4", children: "Use this column as ID" });
        }
        if (!selectedPropertyKey_0 || !selectedProperty) {
          return /* @__PURE__ */ jsxRuntime.jsx(ui.Typography, { variant: "body2", color: "disabled", className: "p-4", children: "Do not import this property" });
        }
        return /* @__PURE__ */ jsxRuntime.jsx(PropertySelectEntry, { propertyKey: selectedPropertyKey_0, property: selectedProperty });
      };
      $[3] = selectedProperty;
      $[4] = t2;
    } else {
      t2 = $[4];
    }
    const renderValue = t2;
    let t3;
    if ($[5] !== onIdSelected || $[6] !== onPropertySelected) {
      t3 = (value) => {
        if (value === internalIDValue) {
          onIdSelected();
          onPropertySelected(null);
        } else {
          if (value === "__do_not_import") {
            onPropertySelected(null);
          } else {
            onPropertySelected(value);
          }
        }
      };
      $[5] = onIdSelected;
      $[6] = onPropertySelected;
      $[7] = t3;
    } else {
      t3 = $[7];
    }
    const onSelectValueChange = t3;
    const t4 = isIdColumn ? internalIDValue : selectedPropertyKey ?? void 0;
    let t5;
    if ($[8] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
      t5 = /* @__PURE__ */ jsxRuntime.jsx(ui.SelectItem, { value: "__do_not_import", children: /* @__PURE__ */ jsxRuntime.jsx(ui.Typography, { variant: "body2", color: "disabled", className: "p-4", children: "Do not import this property" }) });
      $[8] = t5;
    } else {
      t5 = $[8];
    }
    let t6;
    if ($[9] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
      t6 = /* @__PURE__ */ jsxRuntime.jsx(ui.SelectItem, { value: internalIDValue, children: /* @__PURE__ */ jsxRuntime.jsx(ui.Typography, { variant: "body2", className: "p-4", children: "Use this column as ID" }) });
      $[9] = t6;
    } else {
      t6 = $[9];
    }
    let t7;
    if ($[10] !== propertiesAndLevel) {
      t7 = propertiesAndLevel.map(_temp3);
      $[10] = propertiesAndLevel;
      $[11] = t7;
    } else {
      t7 = $[11];
    }
    let t8;
    if ($[12] !== onSelectValueChange || $[13] !== renderValue || $[14] !== t4 || $[15] !== t7) {
      t8 = /* @__PURE__ */ jsxRuntime.jsxs(ui.Select, { value: t4, size: "large", fullWidth: true, onValueChange: onSelectValueChange, renderValue, children: [
        t5,
        t6,
        t7
      ] });
      $[12] = onSelectValueChange;
      $[13] = renderValue;
      $[14] = t4;
      $[15] = t7;
      $[16] = t8;
    } else {
      t8 = $[16];
    }
    return t8;
  }
  function _temp3(t0) {
    const {
      property,
      level,
      propertyKey
    } = t0;
    return /* @__PURE__ */ jsxRuntime.jsx(ui.SelectItem, { value: propertyKey, disabled: property.dataType === "map", children: /* @__PURE__ */ jsxRuntime.jsx(PropertySelectEntry, { propertyKey, property, level }) }, propertyKey);
  }
  function getPropertiesAndLevel(key, property, level) {
    const properties = [];
    properties.push({
      property,
      level,
      propertyKey: key
    });
    if (property.dataType === "map" && property.properties) {
      Object.entries(property.properties).forEach(([childKey, value]) => {
        properties.push(...getPropertiesAndLevel(`${key}.${childKey}`, value, level + 1));
      });
    }
    return properties;
  }
  function PropertySelectEntry(t0) {
    const $ = reactCompilerRuntime.c(21);
    const {
      propertyKey,
      property,
      level: t1
    } = t0;
    const level = t1 === void 0 ? 0 : t1;
    const {
      propertyConfigs
    } = core.useCustomizationController();
    let t2;
    if ($[0] !== property || $[1] !== propertyConfigs) {
      t2 = core.getFieldConfig(property, propertyConfigs);
      $[0] = property;
      $[1] = propertyConfigs;
      $[2] = t2;
    } else {
      t2 = $[2];
    }
    const widget = t2;
    let t3;
    if ($[3] !== level) {
      t3 = new Array(level).fill(0).map(_temp4);
      $[3] = level;
      $[4] = t3;
    } else {
      t3 = $[4];
    }
    const t4 = widget?.name;
    let t5;
    if ($[5] !== widget) {
      t5 = /* @__PURE__ */ jsxRuntime.jsx(core.PropertyConfigBadge, { propertyConfig: widget });
      $[5] = widget;
      $[6] = t5;
    } else {
      t5 = $[6];
    }
    let t6;
    if ($[7] !== t4 || $[8] !== t5) {
      t6 = /* @__PURE__ */ jsxRuntime.jsx("div", { className: "m-4", children: /* @__PURE__ */ jsxRuntime.jsx(ui.Tooltip, { title: t4, children: t5 }) });
      $[7] = t4;
      $[8] = t5;
      $[9] = t6;
    } else {
      t6 = $[9];
    }
    const t7 = property.name ? property.name : " ";
    let t8;
    if ($[10] !== t7) {
      t8 = /* @__PURE__ */ jsxRuntime.jsx(ui.Typography, { variant: "body1", component: "span", className: "flex-grow pr-2", children: t7 });
      $[10] = t7;
      $[11] = t8;
    } else {
      t8 = $[11];
    }
    let t9;
    if ($[12] !== propertyKey) {
      t9 = /* @__PURE__ */ jsxRuntime.jsx(ui.Typography, { className: " pr-2", variant: "body2", component: "span", color: "secondary", children: propertyKey });
      $[12] = propertyKey;
      $[13] = t9;
    } else {
      t9 = $[13];
    }
    let t10;
    if ($[14] !== t8 || $[15] !== t9) {
      t10 = /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "flex flex-col flex-grow p-2 pl-2", children: [
        t8,
        t9
      ] });
      $[14] = t8;
      $[15] = t9;
      $[16] = t10;
    } else {
      t10 = $[16];
    }
    let t11;
    if ($[17] !== t10 || $[18] !== t3 || $[19] !== t6) {
      t11 = /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "flex flex-row w-full text-start items-center h-full", children: [
        t3,
        t6,
        t10
      ] });
      $[17] = t10;
      $[18] = t3;
      $[19] = t6;
      $[20] = t11;
    } else {
      t11 = $[20];
    }
    return t11;
  }
  function _temp4(_, index) {
    return /* @__PURE__ */ jsxRuntime.jsx("div", { className: ui.cls(ui.defaultBorderMixin, "ml-8 border-l h-12") }, index);
  }
  function ImportDataPreview(t0) {
    const $ = reactCompilerRuntime.c(14);
    const {
      importConfig,
      properties
    } = t0;
    const authController = core.useAuthController();
    const navigation = core.useNavigationController();
    let t1;
    if ($[0] !== authController || $[1] !== importConfig || $[2] !== navigation || $[3] !== properties) {
      t1 = () => {
        const mappedData = importConfig.importData.map((d) => convertDataToEntity(authController, navigation, d, importConfig.idColumn, importConfig.headersMapping, properties, "TEMP_PATH", importConfig.defaultValues));
        importConfig.setEntities(mappedData);
      };
      $[0] = authController;
      $[1] = importConfig;
      $[2] = navigation;
      $[3] = properties;
      $[4] = t1;
    } else {
      t1 = $[4];
    }
    let t2;
    if ($[5] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
      t2 = [];
      $[5] = t2;
    } else {
      t2 = $[5];
    }
    React.useEffect(t1, t2);
    const selectionController = core.useSelectionController();
    let t3;
    if ($[6] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
      t3 = /* @__PURE__ */ jsxRuntime.jsxs("div", { children: [
        /* @__PURE__ */ jsxRuntime.jsx(ui.Typography, { variant: "subtitle2", children: "Imported data preview" }),
        /* @__PURE__ */ jsxRuntime.jsx(ui.Typography, { variant: "caption", children: "Entities with the same id will be overwritten" })
      ] });
      $[6] = t3;
    } else {
      t3 = $[6];
    }
    let t4;
    if ($[7] !== importConfig.entities) {
      t4 = {
        data: importConfig.entities,
        dataLoading: false,
        noMoreToLoad: false
      };
      $[7] = importConfig.entities;
      $[8] = t4;
    } else {
      t4 = $[8];
    }
    let t5;
    if ($[9] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
      t5 = /* @__PURE__ */ jsxRuntime.jsx("div", { className: "h-12" });
      $[9] = t5;
    } else {
      t5 = $[9];
    }
    let t6;
    if ($[10] !== properties || $[11] !== selectionController || $[12] !== t4) {
      t6 = /* @__PURE__ */ jsxRuntime.jsx(core.EntityCollectionTable, { title: t3, tableController: t4, enablePopupIcon: false, endAdornment: t5, filterable: false, sortable: false, openEntityMode: "full_screen", selectionController, properties });
      $[10] = properties;
      $[11] = selectionController;
      $[12] = t4;
      $[13] = t6;
    } else {
      t6 = $[13];
    }
    return t6;
  }
  function buildHeadersMappingFromData(objArr, properties) {
    const headersMapping = {};
    objArr.filter(Boolean).forEach((obj) => {
      Object.keys(obj).forEach((key) => {
        const child = obj[key];
        if (typeof child === "object" && !Array.isArray(child)) {
          const childProperty = properties?.[key];
          const childProperties = childProperty && "properties" in childProperty ? childProperty.properties : void 0;
          const childHeadersMapping = buildHeadersMappingFromData([child], childProperties);
          Object.entries(childHeadersMapping).forEach(([subKey, mapping]) => {
            headersMapping[`${key}.${subKey}`] = `${key}.${mapping}`;
          });
        }
        if (!properties) {
          headersMapping[key] = key;
        } else if (key in properties) {
          headersMapping[key] = key;
        } else {
          const slug = core.slugify(key);
          if (slug in properties) {
            headersMapping[key] = slug;
          } else {
            headersMapping[key] = key;
          }
        }
      });
    });
    return headersMapping;
  }
  function useImportPlugin(props) {
    const $ = reactCompilerRuntime.c(3);
    let t0;
    let t1;
    if ($[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
      t1 = [ImportCollectionAction];
      $[0] = t1;
    } else {
      t1 = $[0];
    }
    let t2;
    if ($[1] !== props) {
      t2 = {
        key: "import",
        collectionView: {
          CollectionActions: t1,
          collectionActionsProps: props
        }
      };
      $[1] = props;
      $[2] = t2;
    } else {
      t2 = $[2];
    }
    t0 = t2;
    return t0;
  }
  exports2.DataNewPropertiesMapping = DataNewPropertiesMapping;
  exports2.ImportCollectionAction = ImportCollectionAction;
  exports2.ImportDataPreview = ImportDataPreview;
  exports2.ImportFileUpload = ImportFileUpload;
  exports2.ImportNewPropertyFieldPreview = ImportNewPropertyFieldPreview;
  exports2.ImportSaveInProgress = ImportSaveInProgress;
  exports2.PropertySelectEntry = PropertySelectEntry;
  exports2.convertDataToEntity = convertDataToEntity;
  exports2.convertFileToJson = convertFileToJson;
  exports2.flattenEntry = flattenEntry;
  exports2.getInferenceType = getInferenceType;
  exports2.processValueMapping = processValueMapping;
  exports2.unflattenObject = unflattenObject;
  exports2.useImportConfig = useImportConfig;
  exports2.useImportPlugin = useImportPlugin;
  Object.defineProperty(exports2, Symbol.toStringTag, { value: "Module" });
}));
