// @firecms/data_export/dist/index.es.js
import { c } from "react-compiler-runtime";
import { getArrayValuesCount, getValueInPath, useCustomizationController, useAuthController, useFireCMSContext, useDataSource, useNavigationController, resolveCollection, getDefaultValuesFor } from "@firecms/core";
import { jsxs, Fragment, jsx } from "react/jsx-runtime";
import React, { useCallback } from "react";
import { Tooltip, IconButton, DownloadIcon, Dialog, DialogTitle, DialogContent, cls, BooleanSwitchWithLabel, DialogActions, Button, Alert, CircularProgress } from "@firecms/ui";
function downloadEntitiesExport({
  data,
  additionalData,
  properties,
  propertiesOrder,
  name,
  flattenArrays,
  additionalHeaders,
  exportType,
  dateExportType
}) {
  console.debug("Downloading export", {
    dataLength: data.length,
    properties,
    exportType,
    dateExportType
  });
  if (exportType === "csv") {
    const arrayValuesCount = flattenArrays ? getArrayValuesCount(data.map((d) => d.values)) : {};
    const headers = getExportHeaders(properties, propertiesOrder, additionalHeaders, arrayValuesCount);
    const exportableData = getEntityCSVExportableData(data, additionalData, properties, headers, dateExportType);
    const headersData = entryToCSVRow(headers.map((h) => h.label));
    const csvData = exportableData.map((entry) => entryToCSVRow(entry));
    downloadBlob([headersData, ...csvData], `${name}.csv`, "text/csv");
  } else {
    const exportableData = getEntityJsonExportableData(data, additionalData, properties, dateExportType);
    const json = JSON.stringify(exportableData, null, 2);
    downloadBlob([json], `${name}.json`, "application/json");
  }
}
function getEntityCSVExportableData(data, additionalData, properties, headers, dateExportType) {
  const mergedData = data.map((e) => ({
    id: e.id,
    ...processValuesForExport(e.values, properties, "csv", dateExportType)
  }));
  if (additionalData) {
    additionalData.forEach((additional, index) => {
      mergedData[index] = {
        ...mergedData[index],
        ...additional
      };
    });
  }
  return mergedData && mergedData.map((entry) => {
    return headers.map((header) => getValueInPath(entry, header.key));
  });
}
function getEntityJsonExportableData(data, additionalData, properties, dateExportType) {
  const mergedData = data.map((e) => ({
    id: e.id,
    ...processValuesForExport(e.values, properties, "json", dateExportType)
  }));
  if (additionalData) {
    additionalData.forEach((additional, index) => {
      mergedData[index] = {
        ...mergedData[index],
        ...additional
      };
    });
  }
  return mergedData;
}
function getExportHeaders(properties, propertiesOrder, additionalHeaders, arrayValuesCount) {
  const headers = [{
    label: "id",
    key: "id"
  }, ...(propertiesOrder ?? Object.keys(properties)).flatMap((childKey) => {
    const property = properties[childKey];
    if (!property) {
      console.warn("Property not found", childKey, properties);
      return [];
    }
    if (arrayValuesCount && arrayValuesCount[childKey] > 1) {
      return Array.from({
        length: arrayValuesCount[childKey]
      }, (_, i) => getHeaders(property, `${childKey}[${i}]`, "")).flat();
    } else {
      return getHeaders(property, childKey, "");
    }
  })];
  if (additionalHeaders) {
    headers.push(...additionalHeaders.map((h) => ({
      label: h,
      key: h
    })));
  }
  return headers;
}
function getHeaders(property, propertyKey, prefix = "") {
  const currentKey = prefix ? `${prefix}.${propertyKey}` : propertyKey;
  if (property.dataType === "map" && property.properties) {
    return Object.entries(property.properties).map(([childKey, p]) => getHeaders(p, childKey, currentKey)).flat();
  } else {
    return [{
      label: currentKey,
      key: currentKey
    }];
  }
}
function processValueForExport(inputValue, property, exportType, dateExportType) {
  let value;
  if (property.dataType === "map" && property.properties) {
    value = processValuesForExport(inputValue, property.properties, exportType, dateExportType);
  } else if (property.dataType === "array") {
    if (property.of && Array.isArray(inputValue)) {
      if (Array.isArray(property.of)) {
        value = property.of.map((p, i) => processValueForExport(inputValue[i], p, exportType, dateExportType));
      } else if (property.of.dataType === "map") {
        value = exportType === "csv" ? inputValue.map((e) => JSON.stringify(e)) : inputValue.map((e) => processValueForExport(e, property.of, exportType, dateExportType));
      } else {
        value = inputValue.map((e) => processValueForExport(e, property.of, exportType, dateExportType));
      }
    } else {
      value = inputValue;
    }
  } else if (property.dataType === "reference" && inputValue && inputValue.isEntityReference && inputValue.isEntityReference()) {
    const ref = inputValue ? inputValue : void 0;
    value = ref ? ref.pathWithIdAndDatabase : null;
  } else if (property.dataType === "date" && inputValue instanceof Date) {
    value = inputValue ? dateExportType === "timestamp" ? inputValue.getTime() : inputValue.toISOString() : null;
  } else {
    value = inputValue;
  }
  return value;
}
function processValuesForExport(inputValues, properties, exportType, dateExportType) {
  const updatedValues = Object.entries(properties).map(([key, property]) => {
    const inputValue = inputValues && inputValues[key];
    const updatedValue = processValueForExport(inputValue, property, exportType, dateExportType);
    if (updatedValue === void 0) return {};
    return {
      [key]: updatedValue
    };
  }).reduce((a, b) => ({
    ...a,
    ...b
  }), {});
  return {
    ...inputValues,
    ...updatedValues
  };
}
function entryToCSVRow(entry) {
  return entry.map((v) => {
    if (v === null || v === void 0) return "";
    if (Array.isArray(v)) return '"' + JSON.stringify(v).replaceAll('"', '""') + '"';
    const s = String(v);
    return '"' + s.replaceAll('"', '""') + '"';
  }).join(",") + "\r\n";
}
function downloadBlob(content, filename, contentType) {
  const blob = new Blob(content, {
    type: contentType
  });
  const url = URL.createObjectURL(blob);
  const pom = document.createElement("a");
  pom.href = url;
  pom.setAttribute("download", filename);
  pom.click();
}
function downloadDataAsCsv(data, name) {
  if (data.length === 0) return;
  const headers = Object.keys(data[0]);
  const csvContent = [headers.join(","), ...data.map((row) => headers.map((header) => {
    const value = row[header];
    if (value === null || value === void 0) return "";
    if (Array.isArray(value)) return `"${JSON.stringify(value).replace(/"/g, '""')}"`;
    return `"${String(value).replace(/"/g, '""')}"`;
  }).join(","))].join("\r\n");
  downloadBlob([csvContent], `${name}.csv`, "text/csv");
}
function BasicExportAction({
  data,
  properties,
  propertiesOrder
}) {
  const dateRef = React.useRef(/* @__PURE__ */ new Date());
  const [flattenArrays, setFlattenArrays] = React.useState(true);
  const [exportType, setExportType] = React.useState("csv");
  const [dateExportType, setDateExportType] = React.useState("string");
  const [open, setOpen] = React.useState(false);
  const handleClickOpen = useCallback(() => {
    setOpen(true);
  }, [setOpen]);
  const handleClose = useCallback(() => {
    setOpen(false);
  }, [setOpen]);
  const onOkClicked = useCallback(() => {
    downloadEntitiesExport({
      data,
      additionalData: [],
      properties,
      propertiesOrder,
      name: "export.csv",
      flattenArrays,
      additionalHeaders: [],
      exportType,
      dateExportType
    });
    handleClose();
  }, []);
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(Tooltip, { title: "Export", asChild: true, children: /* @__PURE__ */ jsx(IconButton, { size: "small", color: "primary", onClick: handleClickOpen, children: /* @__PURE__ */ jsx(DownloadIcon, { size: "small" }) }) }),
    /* @__PURE__ */ jsxs(Dialog, { open, onOpenChange: setOpen, maxWidth: "xl", children: [
      /* @__PURE__ */ jsx(DialogTitle, { variant: "h6", children: "Export data" }),
      /* @__PURE__ */ jsxs(DialogContent, { className: "flex flex-col gap-4 my-4", children: [
        /* @__PURE__ */ jsx("div", { children: "Download the the content of this table as a CSV" }),
        /* @__PURE__ */ jsxs("div", { className: "flex flex-row gap-4", children: [
          /* @__PURE__ */ jsxs("div", { className: "p-4 flex flex-col", children: [
            /* @__PURE__ */ jsxs("div", { className: "flex items-center", children: [
              /* @__PURE__ */ jsx("input", { id: "radio-csv", type: "radio", value: "csv", name: "exportType", checked: exportType === "csv", onChange: () => setExportType("csv"), className: cls("w-4 bg-surface-100 border-surface-300 dark:bg-surface-700 dark:border-surface-600") }),
              /* @__PURE__ */ jsx("label", { htmlFor: "radio-csv", className: "p-2 text-sm font-medium text-surface-900 dark:text-surface-accent-300", children: "CSV" })
            ] }),
            /* @__PURE__ */ jsxs("div", { className: "flex items-center", children: [
              /* @__PURE__ */ jsx("input", { id: "radio-json", type: "radio", value: "json", name: "exportType", checked: exportType === "json", onChange: () => setExportType("json"), className: cls("w-4 bg-surface-100 border-surface-300 dark:bg-surface-700 dark:border-surface-600") }),
              /* @__PURE__ */ jsx("label", { htmlFor: "radio-json", className: "p-2 text-sm font-medium text-surface-900 dark:text-surface-accent-300", children: "JSON" })
            ] })
          ] }),
          /* @__PURE__ */ jsxs("div", { className: "p-4 flex flex-col", children: [
            /* @__PURE__ */ jsxs("div", { className: "flex items-center", children: [
              /* @__PURE__ */ jsx("input", { id: "radio-timestamp", type: "radio", value: "timestamp", name: "dateExportType", checked: dateExportType === "timestamp", onChange: () => setDateExportType("timestamp"), className: cls("w-4 bg-surface-100 border-surface-300 dark:bg-surface-700 dark:border-surface-600") }),
              /* @__PURE__ */ jsxs("label", { htmlFor: "radio-timestamp", className: "p-2 text-sm font-medium text-surface-900 dark:text-surface-accent-300", children: [
                "Dates as timestamps (",
                dateRef.current.getTime(),
                ")"
              ] })
            ] }),
            /* @__PURE__ */ jsxs("div", { className: "flex items-center", children: [
              /* @__PURE__ */ jsx("input", { id: "radio-string", type: "radio", value: "string", name: "dateExportType", checked: dateExportType === "string", onChange: () => setDateExportType("string"), className: cls("w-4 bg-surface-100 border-surface-300 dark:bg-surface-700 dark:border-surface-600") }),
              /* @__PURE__ */ jsxs("label", { htmlFor: "radio-string", className: "p-2 text-sm font-medium text-surface-900 dark:text-surface-accent-300", children: [
                "Dates as strings (",
                dateRef.current.toISOString(),
                ")"
              ] })
            ] })
          ] })
        ] }),
        /* @__PURE__ */ jsx(BooleanSwitchWithLabel, { size: "small", disabled: exportType !== "csv", value: flattenArrays, onValueChange: setFlattenArrays, label: "Flatten arrays" })
      ] }),
      /* @__PURE__ */ jsxs(DialogActions, { children: [
        /* @__PURE__ */ jsx(Button, { onClick: handleClose, variant: "text", children: "Cancel" }),
        /* @__PURE__ */ jsx(Button, { onClick: onOkClicked, children: "Download" })
      ] })
    ] })
  ] });
}
const DOCS_LIMIT = 500;
function ExportCollectionAction({
  collection: inputCollection,
  path: inputPath,
  collectionEntitiesCount,
  onAnalyticsEvent,
  exportAllowed,
  notAllowedView
}) {
  const customizationController = useCustomizationController();
  const exportConfig = typeof inputCollection.exportable === "object" ? inputCollection.exportable : void 0;
  const dateRef = React.useRef(/* @__PURE__ */ new Date());
  const [includeUndefinedValues, setIncludeUndefinedValues] = React.useState(false);
  const [flattenArrays, setFlattenArrays] = React.useState(true);
  const [exportType, setExportType] = React.useState("csv");
  const [dateExportType, setDateExportType] = React.useState("string");
  const authController = useAuthController();
  const context = useFireCMSContext();
  const dataSource = useDataSource();
  const navigationController = useNavigationController();
  const path = navigationController.resolveIdsFrom(inputPath);
  const canExport = !exportAllowed || exportAllowed({
    collectionEntitiesCount: collectionEntitiesCount ?? 0,
    path,
    collection: inputCollection
  });
  const collection = React.useMemo(() => resolveCollection({
    collection: inputCollection,
    path,
    propertyConfigs: customizationController.propertyConfigs,
    authController
  }), [inputCollection, path]);
  const [dataLoading, setDataLoading] = React.useState(false);
  const [dataLoadingError, setDataLoadingError] = React.useState();
  const [open, setOpen] = React.useState(false);
  const handleClickOpen = useCallback(() => {
    setOpen(true);
  }, [setOpen]);
  const handleClose = useCallback(() => {
    setOpen(false);
  }, [setOpen]);
  const fetchAdditionalFields = useCallback(async (entities) => {
    const additionalExportFields = exportConfig?.additionalFields;
    const additionalFields = collection.additionalFields;
    const resolvedExportColumnsValues = additionalExportFields ? await Promise.all(entities.map(async (entity) => {
      return (await Promise.all(additionalExportFields.map(async (column) => {
        return {
          [column.key]: await column.builder({
            entity,
            context
          })
        };
      }))).reduce((a, b) => ({
        ...a,
        ...b
      }), {});
    })) : [];
    const resolvedColumnsValues = additionalFields ? await Promise.all(entities.map(async (entity_0) => {
      return (await Promise.all(additionalFields.map(async (field) => {
        if (!field.value) return {};
        return {
          [field.key]: await field.value({
            entity: entity_0,
            context
          })
        };
      }))).reduce((a_0, b_0) => ({
        ...a_0,
        ...b_0
      }), {});
    })) : [];
    return [...resolvedExportColumnsValues, ...resolvedColumnsValues];
  }, [exportConfig?.additionalFields]);
  const doDownload = useCallback(async (collection_0, exportConfig_0) => {
    onAnalyticsEvent?.("export_collection", {
      collection: collection_0.path
    });
    setDataLoading(true);
    dataSource.fetchCollection({
      path,
      collection: collection_0
    }).then(async (data) => {
      setDataLoadingError(void 0);
      const additionalData = await fetchAdditionalFields(data);
      const additionalHeaders = [...exportConfig_0?.additionalFields?.map((column_0) => column_0.key) ?? [], ...collection_0.additionalFields?.map((field_0) => field_0.key) ?? []];
      const dataWithDefaults = includeUndefinedValues ? data.map((entity_1) => {
        const defaultValues = getDefaultValuesFor(collection_0.properties);
        return {
          ...entity_1,
          values: {
            ...defaultValues,
            ...entity_1.values
          }
        };
      }) : data;
      downloadEntitiesExport({
        data: dataWithDefaults,
        additionalData,
        properties: collection_0.properties,
        propertiesOrder: collection_0.propertiesOrder,
        name: collection_0.name,
        flattenArrays,
        additionalHeaders,
        exportType,
        dateExportType
      });
      onAnalyticsEvent?.("export_collection_success", {
        collection: collection_0.path
      });
    }).catch((e) => {
      console.error("Error loading export data", e);
      setDataLoadingError(e);
    }).finally(() => setDataLoading(false));
  }, [onAnalyticsEvent, dataSource, path, fetchAdditionalFields, includeUndefinedValues, flattenArrays, exportType, dateExportType]);
  const onOkClicked = useCallback(() => {
    doDownload(collection, exportConfig);
    handleClose();
  }, [doDownload, collection, exportConfig, handleClose]);
  return /* @__PURE__ */ jsxs(Fragment, { children: [
    /* @__PURE__ */ jsx(Tooltip, { title: "Export", asChild: true, children: /* @__PURE__ */ jsx(IconButton, { size: "small", color: "primary", onClick: handleClickOpen, children: /* @__PURE__ */ jsx(DownloadIcon, { size: "small" }) }) }),
    /* @__PURE__ */ jsxs(Dialog, { open, onOpenChange: setOpen, maxWidth: "xl", children: [
      /* @__PURE__ */ jsx(DialogTitle, { variant: "h6", children: "Export data" }),
      /* @__PURE__ */ jsxs(DialogContent, { className: "flex flex-col gap-4 my-4", children: [
        /* @__PURE__ */ jsx("div", { children: "Download the the content of this table as a CSV" }),
        collectionEntitiesCount !== void 0 && collectionEntitiesCount > DOCS_LIMIT && /* @__PURE__ */ jsx(Alert, { color: "warning", children: /* @__PURE__ */ jsxs("div", { children: [
          "This collections has a large number of documents (",
          collectionEntitiesCount,
          ")."
        ] }) }),
        /* @__PURE__ */ jsxs("div", { className: "flex flex-row gap-4", children: [
          /* @__PURE__ */ jsxs("div", { className: "p-4 flex flex-col", children: [
            /* @__PURE__ */ jsxs("div", { className: "flex items-center", children: [
              /* @__PURE__ */ jsx("input", { id: "radio-csv", type: "radio", value: "csv", name: "exportType", checked: exportType === "csv", onChange: () => setExportType("csv"), className: cls("w-4 bg-surface-100 border-surface-300 dark:bg-surface-700 dark:border-surface-600") }),
              /* @__PURE__ */ jsx("label", { htmlFor: "radio-csv", className: "p-2 text-sm font-medium text-surface-900 dark:text-surface-300", children: "CSV" })
            ] }),
            /* @__PURE__ */ jsxs("div", { className: "flex items-center", children: [
              /* @__PURE__ */ jsx("input", { id: "radio-json", type: "radio", value: "json", name: "exportType", checked: exportType === "json", onChange: () => setExportType("json"), className: cls("w-4 bg-surface-100 border-surface-300 dark:bg-surface-700 dark:border-surface-600") }),
              /* @__PURE__ */ jsx("label", { htmlFor: "radio-json", className: "p-2 text-sm font-medium text-surface-900 dark:text-surface-300", children: "JSON" })
            ] })
          ] }),
          /* @__PURE__ */ jsxs("div", { className: "p-4 flex flex-col", children: [
            /* @__PURE__ */ jsxs("div", { className: "flex items-center", children: [
              /* @__PURE__ */ jsx("input", { id: "radio-timestamp", type: "radio", value: "timestamp", name: "dateExportType", checked: dateExportType === "timestamp", onChange: () => setDateExportType("timestamp"), className: cls("w-4 bg-surface-100 border-surface-300 dark:bg-surface-700 dark:border-surface-600") }),
              /* @__PURE__ */ jsxs("label", { htmlFor: "radio-timestamp", className: "p-2 text-sm font-medium text-surface-900 dark:text-surface-300", children: [
                "Dates as timestamps (",
                dateRef.current.getTime(),
                ")"
              ] })
            ] }),
            /* @__PURE__ */ jsxs("div", { className: "flex items-center", children: [
              /* @__PURE__ */ jsx("input", { id: "radio-string", type: "radio", value: "string", name: "dateExportType", checked: dateExportType === "string", onChange: () => setDateExportType("string"), className: cls("w-4 bg-surface-100 border-surface-300 dark:bg-surface-700 dark:border-surface-600") }),
              /* @__PURE__ */ jsxs("label", { htmlFor: "radio-string", className: "p-2 text-sm font-medium text-surface-900 dark:text-surface-300", children: [
                "Dates as strings (",
                dateRef.current.toISOString(),
                ")"
              ] })
            ] })
          ] })
        ] }),
        /* @__PURE__ */ jsx(BooleanSwitchWithLabel, { size: "small", disabled: exportType !== "csv", value: flattenArrays, onValueChange: setFlattenArrays, label: "Flatten arrays" }),
        /* @__PURE__ */ jsx(BooleanSwitchWithLabel, { size: "small", value: includeUndefinedValues, onValueChange: setIncludeUndefinedValues, label: "Include undefined values" }),
        !canExport && notAllowedView
      ] }),
      /* @__PURE__ */ jsxs(DialogActions, { children: [
        dataLoading && /* @__PURE__ */ jsx(CircularProgress, { size: "smallest" }),
        /* @__PURE__ */ jsx(Button, { onClick: handleClose, variant: "text", children: "Cancel" }),
        /* @__PURE__ */ jsx(Button, { onClick: onOkClicked, disabled: dataLoading || !canExport, children: "Download" })
      ] })
    ] })
  ] });
}
function useExportPlugin(props) {
  const $ = c(3);
  let t0;
  let t1;
  if ($[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
    t1 = [ExportCollectionAction];
    $[0] = t1;
  } else {
    t1 = $[0];
  }
  let t2;
  if ($[1] !== props) {
    t2 = {
      key: "export",
      collectionView: {
        CollectionActions: t1,
        collectionActionsProps: props
      }
    };
    $[1] = props;
    $[2] = t2;
  } else {
    t2 = $[2];
  }
  t0 = t2;
  return t0;
}
export {
  BasicExportAction,
  ExportCollectionAction,
  downloadBlob,
  downloadDataAsCsv,
  downloadEntitiesExport,
  getEntityCSVExportableData,
  getEntityJsonExportableData,
  useExportPlugin
};
