// @firecms/data_export/dist/index.umd.js
(function(global, factory) {
  typeof exports === "object" && typeof module !== "undefined" ? factory(exports, require("react-compiler-runtime"), require("@firecms/core"), require("react/jsx-runtime"), require("react"), require("@firecms/ui")) : typeof define === "function" && define.amd ? define(["exports", "react-compiler-runtime", "@firecms/core", "react/jsx-runtime", "react", "@firecms/ui"], factory) : (global = typeof globalThis !== "undefined" ? globalThis : global || self, factory(global["FireCMS data import/export"] = {}, global.reactCompilerRuntime, global.core, global.jsxRuntime, global.React, global.ui));
})(this, (function(exports2, reactCompilerRuntime, core, jsxRuntime, React, ui) {
  "use strict";
  function downloadEntitiesExport({
    data,
    additionalData,
    properties,
    propertiesOrder,
    name,
    flattenArrays,
    additionalHeaders,
    exportType,
    dateExportType
  }) {
    console.debug("Downloading export", {
      dataLength: data.length,
      properties,
      exportType,
      dateExportType
    });
    if (exportType === "csv") {
      const arrayValuesCount = flattenArrays ? core.getArrayValuesCount(data.map((d) => d.values)) : {};
      const headers = getExportHeaders(properties, propertiesOrder, additionalHeaders, arrayValuesCount);
      const exportableData = getEntityCSVExportableData(data, additionalData, properties, headers, dateExportType);
      const headersData = entryToCSVRow(headers.map((h) => h.label));
      const csvData = exportableData.map((entry) => entryToCSVRow(entry));
      downloadBlob([headersData, ...csvData], `${name}.csv`, "text/csv");
    } else {
      const exportableData = getEntityJsonExportableData(data, additionalData, properties, dateExportType);
      const json = JSON.stringify(exportableData, null, 2);
      downloadBlob([json], `${name}.json`, "application/json");
    }
  }
  function getEntityCSVExportableData(data, additionalData, properties, headers, dateExportType) {
    const mergedData = data.map((e) => ({
      id: e.id,
      ...processValuesForExport(e.values, properties, "csv", dateExportType)
    }));
    if (additionalData) {
      additionalData.forEach((additional, index) => {
        mergedData[index] = {
          ...mergedData[index],
          ...additional
        };
      });
    }
    return mergedData && mergedData.map((entry) => {
      return headers.map((header) => core.getValueInPath(entry, header.key));
    });
  }
  function getEntityJsonExportableData(data, additionalData, properties, dateExportType) {
    const mergedData = data.map((e) => ({
      id: e.id,
      ...processValuesForExport(e.values, properties, "json", dateExportType)
    }));
    if (additionalData) {
      additionalData.forEach((additional, index) => {
        mergedData[index] = {
          ...mergedData[index],
          ...additional
        };
      });
    }
    return mergedData;
  }
  function getExportHeaders(properties, propertiesOrder, additionalHeaders, arrayValuesCount) {
    const headers = [{
      label: "id",
      key: "id"
    }, ...(propertiesOrder ?? Object.keys(properties)).flatMap((childKey) => {
      const property = properties[childKey];
      if (!property) {
        console.warn("Property not found", childKey, properties);
        return [];
      }
      if (arrayValuesCount && arrayValuesCount[childKey] > 1) {
        return Array.from({
          length: arrayValuesCount[childKey]
        }, (_, i) => getHeaders(property, `${childKey}[${i}]`, "")).flat();
      } else {
        return getHeaders(property, childKey, "");
      }
    })];
    if (additionalHeaders) {
      headers.push(...additionalHeaders.map((h) => ({
        label: h,
        key: h
      })));
    }
    return headers;
  }
  function getHeaders(property, propertyKey, prefix = "") {
    const currentKey = prefix ? `${prefix}.${propertyKey}` : propertyKey;
    if (property.dataType === "map" && property.properties) {
      return Object.entries(property.properties).map(([childKey, p]) => getHeaders(p, childKey, currentKey)).flat();
    } else {
      return [{
        label: currentKey,
        key: currentKey
      }];
    }
  }
  function processValueForExport(inputValue, property, exportType, dateExportType) {
    let value;
    if (property.dataType === "map" && property.properties) {
      value = processValuesForExport(inputValue, property.properties, exportType, dateExportType);
    } else if (property.dataType === "array") {
      if (property.of && Array.isArray(inputValue)) {
        if (Array.isArray(property.of)) {
          value = property.of.map((p, i) => processValueForExport(inputValue[i], p, exportType, dateExportType));
        } else if (property.of.dataType === "map") {
          value = exportType === "csv" ? inputValue.map((e) => JSON.stringify(e)) : inputValue.map((e) => processValueForExport(e, property.of, exportType, dateExportType));
        } else {
          value = inputValue.map((e) => processValueForExport(e, property.of, exportType, dateExportType));
        }
      } else {
        value = inputValue;
      }
    } else if (property.dataType === "reference" && inputValue && inputValue.isEntityReference && inputValue.isEntityReference()) {
      const ref = inputValue ? inputValue : void 0;
      value = ref ? ref.pathWithIdAndDatabase : null;
    } else if (property.dataType === "date" && inputValue instanceof Date) {
      value = inputValue ? dateExportType === "timestamp" ? inputValue.getTime() : inputValue.toISOString() : null;
    } else {
      value = inputValue;
    }
    return value;
  }
  function processValuesForExport(inputValues, properties, exportType, dateExportType) {
    const updatedValues = Object.entries(properties).map(([key, property]) => {
      const inputValue = inputValues && inputValues[key];
      const updatedValue = processValueForExport(inputValue, property, exportType, dateExportType);
      if (updatedValue === void 0) return {};
      return {
        [key]: updatedValue
      };
    }).reduce((a, b) => ({
      ...a,
      ...b
    }), {});
    return {
      ...inputValues,
      ...updatedValues
    };
  }
  function entryToCSVRow(entry) {
    return entry.map((v) => {
      if (v === null || v === void 0) return "";
      if (Array.isArray(v)) return '"' + JSON.stringify(v).replaceAll('"', '""') + '"';
      const s = String(v);
      return '"' + s.replaceAll('"', '""') + '"';
    }).join(",") + "\r\n";
  }
  function downloadBlob(content, filename, contentType) {
    const blob = new Blob(content, {
      type: contentType
    });
    const url = URL.createObjectURL(blob);
    const pom = document.createElement("a");
    pom.href = url;
    pom.setAttribute("download", filename);
    pom.click();
  }
  function downloadDataAsCsv(data, name) {
    if (data.length === 0) return;
    const headers = Object.keys(data[0]);
    const csvContent = [headers.join(","), ...data.map((row) => headers.map((header) => {
      const value = row[header];
      if (value === null || value === void 0) return "";
      if (Array.isArray(value)) return `"${JSON.stringify(value).replace(/"/g, '""')}"`;
      return `"${String(value).replace(/"/g, '""')}"`;
    }).join(","))].join("\r\n");
    downloadBlob([csvContent], `${name}.csv`, "text/csv");
  }
  function BasicExportAction({
    data,
    properties,
    propertiesOrder
  }) {
    const dateRef = React.useRef(/* @__PURE__ */ new Date());
    const [flattenArrays, setFlattenArrays] = React.useState(true);
    const [exportType, setExportType] = React.useState("csv");
    const [dateExportType, setDateExportType] = React.useState("string");
    const [open, setOpen] = React.useState(false);
    const handleClickOpen = React.useCallback(() => {
      setOpen(true);
    }, [setOpen]);
    const handleClose = React.useCallback(() => {
      setOpen(false);
    }, [setOpen]);
    const onOkClicked = React.useCallback(() => {
      downloadEntitiesExport({
        data,
        additionalData: [],
        properties,
        propertiesOrder,
        name: "export.csv",
        flattenArrays,
        additionalHeaders: [],
        exportType,
        dateExportType
      });
      handleClose();
    }, []);
    return /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
      /* @__PURE__ */ jsxRuntime.jsx(ui.Tooltip, { title: "Export", asChild: true, children: /* @__PURE__ */ jsxRuntime.jsx(ui.IconButton, { size: "small", color: "primary", onClick: handleClickOpen, children: /* @__PURE__ */ jsxRuntime.jsx(ui.DownloadIcon, { size: "small" }) }) }),
      /* @__PURE__ */ jsxRuntime.jsxs(ui.Dialog, { open, onOpenChange: setOpen, maxWidth: "xl", children: [
        /* @__PURE__ */ jsxRuntime.jsx(ui.DialogTitle, { variant: "h6", children: "Export data" }),
        /* @__PURE__ */ jsxRuntime.jsxs(ui.DialogContent, { className: "flex flex-col gap-4 my-4", children: [
          /* @__PURE__ */ jsxRuntime.jsx("div", { children: "Download the the content of this table as a CSV" }),
          /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "flex flex-row gap-4", children: [
            /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "p-4 flex flex-col", children: [
              /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "flex items-center", children: [
                /* @__PURE__ */ jsxRuntime.jsx("input", { id: "radio-csv", type: "radio", value: "csv", name: "exportType", checked: exportType === "csv", onChange: () => setExportType("csv"), className: ui.cls("w-4 bg-surface-100 border-surface-300 dark:bg-surface-700 dark:border-surface-600") }),
                /* @__PURE__ */ jsxRuntime.jsx("label", { htmlFor: "radio-csv", className: "p-2 text-sm font-medium text-surface-900 dark:text-surface-accent-300", children: "CSV" })
              ] }),
              /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "flex items-center", children: [
                /* @__PURE__ */ jsxRuntime.jsx("input", { id: "radio-json", type: "radio", value: "json", name: "exportType", checked: exportType === "json", onChange: () => setExportType("json"), className: ui.cls("w-4 bg-surface-100 border-surface-300 dark:bg-surface-700 dark:border-surface-600") }),
                /* @__PURE__ */ jsxRuntime.jsx("label", { htmlFor: "radio-json", className: "p-2 text-sm font-medium text-surface-900 dark:text-surface-accent-300", children: "JSON" })
              ] })
            ] }),
            /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "p-4 flex flex-col", children: [
              /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "flex items-center", children: [
                /* @__PURE__ */ jsxRuntime.jsx("input", { id: "radio-timestamp", type: "radio", value: "timestamp", name: "dateExportType", checked: dateExportType === "timestamp", onChange: () => setDateExportType("timestamp"), className: ui.cls("w-4 bg-surface-100 border-surface-300 dark:bg-surface-700 dark:border-surface-600") }),
                /* @__PURE__ */ jsxRuntime.jsxs("label", { htmlFor: "radio-timestamp", className: "p-2 text-sm font-medium text-surface-900 dark:text-surface-accent-300", children: [
                  "Dates as timestamps (",
                  dateRef.current.getTime(),
                  ")"
                ] })
              ] }),
              /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "flex items-center", children: [
                /* @__PURE__ */ jsxRuntime.jsx("input", { id: "radio-string", type: "radio", value: "string", name: "dateExportType", checked: dateExportType === "string", onChange: () => setDateExportType("string"), className: ui.cls("w-4 bg-surface-100 border-surface-300 dark:bg-surface-700 dark:border-surface-600") }),
                /* @__PURE__ */ jsxRuntime.jsxs("label", { htmlFor: "radio-string", className: "p-2 text-sm font-medium text-surface-900 dark:text-surface-accent-300", children: [
                  "Dates as strings (",
                  dateRef.current.toISOString(),
                  ")"
                ] })
              ] })
            ] })
          ] }),
          /* @__PURE__ */ jsxRuntime.jsx(ui.BooleanSwitchWithLabel, { size: "small", disabled: exportType !== "csv", value: flattenArrays, onValueChange: setFlattenArrays, label: "Flatten arrays" })
        ] }),
        /* @__PURE__ */ jsxRuntime.jsxs(ui.DialogActions, { children: [
          /* @__PURE__ */ jsxRuntime.jsx(ui.Button, { onClick: handleClose, variant: "text", children: "Cancel" }),
          /* @__PURE__ */ jsxRuntime.jsx(ui.Button, { onClick: onOkClicked, children: "Download" })
        ] })
      ] })
    ] });
  }
  const DOCS_LIMIT = 500;
  function ExportCollectionAction({
    collection: inputCollection,
    path: inputPath,
    collectionEntitiesCount,
    onAnalyticsEvent,
    exportAllowed,
    notAllowedView
  }) {
    const customizationController = core.useCustomizationController();
    const exportConfig = typeof inputCollection.exportable === "object" ? inputCollection.exportable : void 0;
    const dateRef = React.useRef(/* @__PURE__ */ new Date());
    const [includeUndefinedValues, setIncludeUndefinedValues] = React.useState(false);
    const [flattenArrays, setFlattenArrays] = React.useState(true);
    const [exportType, setExportType] = React.useState("csv");
    const [dateExportType, setDateExportType] = React.useState("string");
    const authController = core.useAuthController();
    const context = core.useFireCMSContext();
    const dataSource = core.useDataSource();
    const navigationController = core.useNavigationController();
    const path = navigationController.resolveIdsFrom(inputPath);
    const canExport = !exportAllowed || exportAllowed({
      collectionEntitiesCount: collectionEntitiesCount ?? 0,
      path,
      collection: inputCollection
    });
    const collection = React.useMemo(() => core.resolveCollection({
      collection: inputCollection,
      path,
      propertyConfigs: customizationController.propertyConfigs,
      authController
    }), [inputCollection, path]);
    const [dataLoading, setDataLoading] = React.useState(false);
    const [dataLoadingError, setDataLoadingError] = React.useState();
    const [open, setOpen] = React.useState(false);
    const handleClickOpen = React.useCallback(() => {
      setOpen(true);
    }, [setOpen]);
    const handleClose = React.useCallback(() => {
      setOpen(false);
    }, [setOpen]);
    const fetchAdditionalFields = React.useCallback(async (entities) => {
      const additionalExportFields = exportConfig?.additionalFields;
      const additionalFields = collection.additionalFields;
      const resolvedExportColumnsValues = additionalExportFields ? await Promise.all(entities.map(async (entity) => {
        return (await Promise.all(additionalExportFields.map(async (column) => {
          return {
            [column.key]: await column.builder({
              entity,
              context
            })
          };
        }))).reduce((a, b) => ({
          ...a,
          ...b
        }), {});
      })) : [];
      const resolvedColumnsValues = additionalFields ? await Promise.all(entities.map(async (entity_0) => {
        return (await Promise.all(additionalFields.map(async (field) => {
          if (!field.value) return {};
          return {
            [field.key]: await field.value({
              entity: entity_0,
              context
            })
          };
        }))).reduce((a_0, b_0) => ({
          ...a_0,
          ...b_0
        }), {});
      })) : [];
      return [...resolvedExportColumnsValues, ...resolvedColumnsValues];
    }, [exportConfig?.additionalFields]);
    const doDownload = React.useCallback(async (collection_0, exportConfig_0) => {
      onAnalyticsEvent?.("export_collection", {
        collection: collection_0.path
      });
      setDataLoading(true);
      dataSource.fetchCollection({
        path,
        collection: collection_0
      }).then(async (data) => {
        setDataLoadingError(void 0);
        const additionalData = await fetchAdditionalFields(data);
        const additionalHeaders = [...exportConfig_0?.additionalFields?.map((column_0) => column_0.key) ?? [], ...collection_0.additionalFields?.map((field_0) => field_0.key) ?? []];
        const dataWithDefaults = includeUndefinedValues ? data.map((entity_1) => {
          const defaultValues = core.getDefaultValuesFor(collection_0.properties);
          return {
            ...entity_1,
            values: {
              ...defaultValues,
              ...entity_1.values
            }
          };
        }) : data;
        downloadEntitiesExport({
          data: dataWithDefaults,
          additionalData,
          properties: collection_0.properties,
          propertiesOrder: collection_0.propertiesOrder,
          name: collection_0.name,
          flattenArrays,
          additionalHeaders,
          exportType,
          dateExportType
        });
        onAnalyticsEvent?.("export_collection_success", {
          collection: collection_0.path
        });
      }).catch((e) => {
        console.error("Error loading export data", e);
        setDataLoadingError(e);
      }).finally(() => setDataLoading(false));
    }, [onAnalyticsEvent, dataSource, path, fetchAdditionalFields, includeUndefinedValues, flattenArrays, exportType, dateExportType]);
    const onOkClicked = React.useCallback(() => {
      doDownload(collection, exportConfig);
      handleClose();
    }, [doDownload, collection, exportConfig, handleClose]);
    return /* @__PURE__ */ jsxRuntime.jsxs(jsxRuntime.Fragment, { children: [
      /* @__PURE__ */ jsxRuntime.jsx(ui.Tooltip, { title: "Export", asChild: true, children: /* @__PURE__ */ jsxRuntime.jsx(ui.IconButton, { size: "small", color: "primary", onClick: handleClickOpen, children: /* @__PURE__ */ jsxRuntime.jsx(ui.DownloadIcon, { size: "small" }) }) }),
      /* @__PURE__ */ jsxRuntime.jsxs(ui.Dialog, { open, onOpenChange: setOpen, maxWidth: "xl", children: [
        /* @__PURE__ */ jsxRuntime.jsx(ui.DialogTitle, { variant: "h6", children: "Export data" }),
        /* @__PURE__ */ jsxRuntime.jsxs(ui.DialogContent, { className: "flex flex-col gap-4 my-4", children: [
          /* @__PURE__ */ jsxRuntime.jsx("div", { children: "Download the the content of this table as a CSV" }),
          collectionEntitiesCount !== void 0 && collectionEntitiesCount > DOCS_LIMIT && /* @__PURE__ */ jsxRuntime.jsx(ui.Alert, { color: "warning", children: /* @__PURE__ */ jsxRuntime.jsxs("div", { children: [
            "This collections has a large number of documents (",
            collectionEntitiesCount,
            ")."
          ] }) }),
          /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "flex flex-row gap-4", children: [
            /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "p-4 flex flex-col", children: [
              /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "flex items-center", children: [
                /* @__PURE__ */ jsxRuntime.jsx("input", { id: "radio-csv", type: "radio", value: "csv", name: "exportType", checked: exportType === "csv", onChange: () => setExportType("csv"), className: ui.cls("w-4 bg-surface-100 border-surface-300 dark:bg-surface-700 dark:border-surface-600") }),
                /* @__PURE__ */ jsxRuntime.jsx("label", { htmlFor: "radio-csv", className: "p-2 text-sm font-medium text-surface-900 dark:text-surface-300", children: "CSV" })
              ] }),
              /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "flex items-center", children: [
                /* @__PURE__ */ jsxRuntime.jsx("input", { id: "radio-json", type: "radio", value: "json", name: "exportType", checked: exportType === "json", onChange: () => setExportType("json"), className: ui.cls("w-4 bg-surface-100 border-surface-300 dark:bg-surface-700 dark:border-surface-600") }),
                /* @__PURE__ */ jsxRuntime.jsx("label", { htmlFor: "radio-json", className: "p-2 text-sm font-medium text-surface-900 dark:text-surface-300", children: "JSON" })
              ] })
            ] }),
            /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "p-4 flex flex-col", children: [
              /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "flex items-center", children: [
                /* @__PURE__ */ jsxRuntime.jsx("input", { id: "radio-timestamp", type: "radio", value: "timestamp", name: "dateExportType", checked: dateExportType === "timestamp", onChange: () => setDateExportType("timestamp"), className: ui.cls("w-4 bg-surface-100 border-surface-300 dark:bg-surface-700 dark:border-surface-600") }),
                /* @__PURE__ */ jsxRuntime.jsxs("label", { htmlFor: "radio-timestamp", className: "p-2 text-sm font-medium text-surface-900 dark:text-surface-300", children: [
                  "Dates as timestamps (",
                  dateRef.current.getTime(),
                  ")"
                ] })
              ] }),
              /* @__PURE__ */ jsxRuntime.jsxs("div", { className: "flex items-center", children: [
                /* @__PURE__ */ jsxRuntime.jsx("input", { id: "radio-string", type: "radio", value: "string", name: "dateExportType", checked: dateExportType === "string", onChange: () => setDateExportType("string"), className: ui.cls("w-4 bg-surface-100 border-surface-300 dark:bg-surface-700 dark:border-surface-600") }),
                /* @__PURE__ */ jsxRuntime.jsxs("label", { htmlFor: "radio-string", className: "p-2 text-sm font-medium text-surface-900 dark:text-surface-300", children: [
                  "Dates as strings (",
                  dateRef.current.toISOString(),
                  ")"
                ] })
              ] })
            ] })
          ] }),
          /* @__PURE__ */ jsxRuntime.jsx(ui.BooleanSwitchWithLabel, { size: "small", disabled: exportType !== "csv", value: flattenArrays, onValueChange: setFlattenArrays, label: "Flatten arrays" }),
          /* @__PURE__ */ jsxRuntime.jsx(ui.BooleanSwitchWithLabel, { size: "small", value: includeUndefinedValues, onValueChange: setIncludeUndefinedValues, label: "Include undefined values" }),
          !canExport && notAllowedView
        ] }),
        /* @__PURE__ */ jsxRuntime.jsxs(ui.DialogActions, { children: [
          dataLoading && /* @__PURE__ */ jsxRuntime.jsx(ui.CircularProgress, { size: "smallest" }),
          /* @__PURE__ */ jsxRuntime.jsx(ui.Button, { onClick: handleClose, variant: "text", children: "Cancel" }),
          /* @__PURE__ */ jsxRuntime.jsx(ui.Button, { onClick: onOkClicked, disabled: dataLoading || !canExport, children: "Download" })
        ] })
      ] })
    ] });
  }
  function useExportPlugin(props) {
    const $ = reactCompilerRuntime.c(3);
    let t0;
    let t1;
    if ($[0] === /* @__PURE__ */ Symbol.for("react.memo_cache_sentinel")) {
      t1 = [ExportCollectionAction];
      $[0] = t1;
    } else {
      t1 = $[0];
    }
    let t2;
    if ($[1] !== props) {
      t2 = {
        key: "export",
        collectionView: {
          CollectionActions: t1,
          collectionActionsProps: props
        }
      };
      $[1] = props;
      $[2] = t2;
    } else {
      t2 = $[2];
    }
    t0 = t2;
    return t0;
  }
  exports2.BasicExportAction = BasicExportAction;
  exports2.ExportCollectionAction = ExportCollectionAction;
  exports2.downloadBlob = downloadBlob;
  exports2.downloadDataAsCsv = downloadDataAsCsv;
  exports2.downloadEntitiesExport = downloadEntitiesExport;
  exports2.getEntityCSVExportableData = getEntityCSVExportableData;
  exports2.getEntityJsonExportableData = getEntityJsonExportableData;
  exports2.useExportPlugin = useExportPlugin;
  Object.defineProperty(exports2, Symbol.toStringTag, { value: "Module" });
}));
